// @synthetixio/core-contracts/typechain-types/mocks/utils/SetUtilMock.sol/AddressSetMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../../common";

export interface AddressSetMockInterface extends utils.Interface {
  functions: {
    "add(address)": FunctionFragment;
    "contains(address)": FunctionFragment;
    "length()": FunctionFragment;
    "positionOf(address)": FunctionFragment;
    "remove(address)": FunctionFragment;
    "replace(address,address)": FunctionFragment;
    "valueAt(uint256)": FunctionFragment;
    "values()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "add"
      | "contains"
      | "length"
      | "positionOf"
      | "remove"
      | "replace"
      | "valueAt"
      | "values"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "add", values: [string]): string;
  encodeFunctionData(functionFragment: "contains", values: [string]): string;
  encodeFunctionData(functionFragment: "length", values?: undefined): string;
  encodeFunctionData(functionFragment: "positionOf", values: [string]): string;
  encodeFunctionData(functionFragment: "remove", values: [string]): string;
  encodeFunctionData(
    functionFragment: "replace",
    values: [string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "valueAt",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(functionFragment: "values", values?: undefined): string;

  decodeFunctionResult(functionFragment: "add", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "contains", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "length", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "positionOf", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "remove", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "replace", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "valueAt", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "values", data: BytesLike): Result;

  events: {};
}

export interface AddressSetMock extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: AddressSetMockInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    add(
      value: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    contains(value: string, overrides?: CallOverrides): Promise<[boolean]>;

    length(overrides?: CallOverrides): Promise<[BigNumber]>;

    positionOf(value: string, overrides?: CallOverrides): Promise<[BigNumber]>;

    remove(
      value: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    replace(
      value: string,
      newValue: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    valueAt(
      position: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[string]>;

    values(overrides?: CallOverrides): Promise<[string[]]>;
  };

  add(
    value: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  contains(value: string, overrides?: CallOverrides): Promise<boolean>;

  length(overrides?: CallOverrides): Promise<BigNumber>;

  positionOf(value: string, overrides?: CallOverrides): Promise<BigNumber>;

  remove(
    value: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  replace(
    value: string,
    newValue: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  valueAt(position: BigNumberish, overrides?: CallOverrides): Promise<string>;

  values(overrides?: CallOverrides): Promise<string[]>;

  callStatic: {
    add(value: string, overrides?: CallOverrides): Promise<void>;

    contains(value: string, overrides?: CallOverrides): Promise<boolean>;

    length(overrides?: CallOverrides): Promise<BigNumber>;

    positionOf(value: string, overrides?: CallOverrides): Promise<BigNumber>;

    remove(value: string, overrides?: CallOverrides): Promise<void>;

    replace(
      value: string,
      newValue: string,
      overrides?: CallOverrides
    ): Promise<void>;

    valueAt(position: BigNumberish, overrides?: CallOverrides): Promise<string>;

    values(overrides?: CallOverrides): Promise<string[]>;
  };

  filters: {};

  estimateGas: {
    add(
      value: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    contains(value: string, overrides?: CallOverrides): Promise<BigNumber>;

    length(overrides?: CallOverrides): Promise<BigNumber>;

    positionOf(value: string, overrides?: CallOverrides): Promise<BigNumber>;

    remove(
      value: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    replace(
      value: string,
      newValue: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    valueAt(
      position: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    values(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    add(
      value: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    contains(
      value: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    length(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    positionOf(
      value: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    remove(
      value: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    replace(
      value: string,
      newValue: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    valueAt(
      position: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    values(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
