// @synthetixio/perps-market/typechain-types/factories/contracts/interfaces/external/IFeeCollector__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IFeeCollector,
  IFeeCollectorInterface,
} from "../../../../contracts/interfaces/external/IFeeCollector";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "feeAmount",
        type: "uint256",
      },
      {
        internalType: "address",
        name: "transactor",
        type: "address",
      },
    ],
    name: "quoteFees",
    outputs: [
      {
        internalType: "uint256",
        name: "feeAmountToCollect",
        type: "uint256",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes4",
        name: "interfaceID",
        type: "bytes4",
      },
    ],
    name: "supportsInterface",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class IFeeCollector__factory {
  static readonly abi = _abi;
  static createInterface(): IFeeCollectorInterface {
    return new utils.Interface(_abi) as IFeeCollectorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IFeeCollector {
    return new Contract(address, _abi, signerOrProvider) as IFeeCollector;
  }
}
