// @synthetixio/perps-market/typechain-types/factories/@synthetixio/oracle-manager/contracts/mocks/pyth/PythErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  PythErrors,
  PythErrorsInterface,
} from "../../../../../../@synthetixio/oracle-manager/contracts/mocks/pyth/PythErrors";

const _abi = [
  {
    inputs: [],
    name: "InsufficientFee",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidArgument",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidGovernanceDataSource",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidGovernanceMessage",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidGovernanceTarget",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidUpdateData",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidUpdateDataSource",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidWormholeAddressToSet",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidWormholeVaa",
    type: "error",
  },
  {
    inputs: [],
    name: "NoFreshUpdate",
    type: "error",
  },
  {
    inputs: [],
    name: "OldGovernanceMessage",
    type: "error",
  },
  {
    inputs: [],
    name: "PriceFeedNotFound",
    type: "error",
  },
  {
    inputs: [],
    name: "PriceFeedNotFoundWithinRange",
    type: "error",
  },
  {
    inputs: [],
    name: "StalePrice",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220381a3641d737e74807735d957b9cc00659a42d48c5795bed155c300497d14f3364736f6c63430008160033";

type PythErrorsConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: PythErrorsConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class PythErrors__factory extends ContractFactory {
  constructor(...args: PythErrorsConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<PythErrors> {
    return super.deploy(overrides || {}) as Promise<PythErrors>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): PythErrors {
    return super.attach(address) as PythErrors;
  }
  override connect(signer: Signer): PythErrors__factory {
    return super.connect(signer) as PythErrors__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): PythErrorsInterface {
    return new utils.Interface(_abi) as PythErrorsInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): PythErrors {
    return new Contract(address, _abi, signerOrProvider) as PythErrors;
  }
}
