// @synthetixio/core-contracts/typechain-types/mocks/utils/AddressUtilMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface AddressUtilMockInterface extends utils.Interface {
  functions: {
    "isContract(address)": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "isContract"): FunctionFragment;

  encodeFunctionData(functionFragment: "isContract", values: [string]): string;

  decodeFunctionResult(functionFragment: "isContract", data: BytesLike): Result;

  events: {};
}

export interface AddressUtilMock extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: AddressUtilMockInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    isContract(account: string, overrides?: CallOverrides): Promise<[boolean]>;
  };

  isContract(account: string, overrides?: CallOverrides): Promise<boolean>;

  callStatic: {
    isContract(account: string, overrides?: CallOverrides): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    isContract(account: string, overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    isContract(
      account: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
