// @synthetixio/core-contracts/typechain-types/mocks/utils/ConfigurableMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface ConfigurableMockInterface extends utils.Interface {
  functions: {
    "acceptConfigurerRole()": FunctionFragment;
    "acceptOwnership()": FunctionFragment;
    "configurer()": FunctionFragment;
    "countUp()": FunctionFragment;
    "counter()": FunctionFragment;
    "nominateNewConfigurer(address)": FunctionFragment;
    "nominateNewOwner(address)": FunctionFragment;
    "nominatedConfigurer()": FunctionFragment;
    "nominatedOwner()": FunctionFragment;
    "owner()": FunctionFragment;
    "renounceConfigurerNomination()": FunctionFragment;
    "renounceNomination()": FunctionFragment;
    "setConfigurer(address)": FunctionFragment;
    "setCounter(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "acceptConfigurerRole"
      | "acceptOwnership"
      | "configurer"
      | "countUp"
      | "counter"
      | "nominateNewConfigurer"
      | "nominateNewOwner"
      | "nominatedConfigurer"
      | "nominatedOwner"
      | "owner"
      | "renounceConfigurerNomination"
      | "renounceNomination"
      | "setConfigurer"
      | "setCounter"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "acceptConfigurerRole",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "acceptOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "configurer",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "countUp", values?: undefined): string;
  encodeFunctionData(functionFragment: "counter", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "nominateNewConfigurer",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "nominateNewOwner",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "nominatedConfigurer",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "nominatedOwner",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "renounceConfigurerNomination",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "renounceNomination",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setConfigurer",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "setCounter",
    values: [BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "acceptConfigurerRole",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "acceptOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "configurer", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "countUp", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "counter", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "nominateNewConfigurer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "nominateNewOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "nominatedConfigurer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "nominatedOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "renounceConfigurerNomination",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceNomination",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setConfigurer",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "setCounter", data: BytesLike): Result;

  events: {
    "ConfigurerChanged(address,address)": EventFragment;
    "ConfigurerNominated(address)": EventFragment;
    "OwnerChanged(address,address)": EventFragment;
    "OwnerNominated(address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "ConfigurerChanged"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "ConfigurerNominated"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "OwnerChanged"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "OwnerNominated"): EventFragment;
}

export interface ConfigurerChangedEventObject {
  oldConfigurer: string;
  newConfigurer: string;
}
export type ConfigurerChangedEvent = TypedEvent<
  [string, string],
  ConfigurerChangedEventObject
>;

export type ConfigurerChangedEventFilter =
  TypedEventFilter<ConfigurerChangedEvent>;

export interface ConfigurerNominatedEventObject {
  newConfigurer: string;
}
export type ConfigurerNominatedEvent = TypedEvent<
  [string],
  ConfigurerNominatedEventObject
>;

export type ConfigurerNominatedEventFilter =
  TypedEventFilter<ConfigurerNominatedEvent>;

export interface OwnerChangedEventObject {
  oldOwner: string;
  newOwner: string;
}
export type OwnerChangedEvent = TypedEvent<
  [string, string],
  OwnerChangedEventObject
>;

export type OwnerChangedEventFilter = TypedEventFilter<OwnerChangedEvent>;

export interface OwnerNominatedEventObject {
  newOwner: string;
}
export type OwnerNominatedEvent = TypedEvent<
  [string],
  OwnerNominatedEventObject
>;

export type OwnerNominatedEventFilter = TypedEventFilter<OwnerNominatedEvent>;

export interface ConfigurableMock extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: ConfigurableMockInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    acceptConfigurerRole(
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    acceptOwnership(
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    configurer(overrides?: CallOverrides): Promise<[string]>;

    countUp(
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    counter(overrides?: CallOverrides): Promise<[BigNumber]>;

    nominateNewConfigurer(
      newNominatedConfigurer: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    nominateNewOwner(
      newNominatedOwner: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    nominatedConfigurer(overrides?: CallOverrides): Promise<[string]>;

    nominatedOwner(overrides?: CallOverrides): Promise<[string]>;

    owner(overrides?: CallOverrides): Promise<[string]>;

    renounceConfigurerNomination(
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    renounceNomination(
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    setConfigurer(
      newConfigurer: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    setCounter(
      _counter: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;
  };

  acceptConfigurerRole(
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  acceptOwnership(
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  configurer(overrides?: CallOverrides): Promise<string>;

  countUp(
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  counter(overrides?: CallOverrides): Promise<BigNumber>;

  nominateNewConfigurer(
    newNominatedConfigurer: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  nominateNewOwner(
    newNominatedOwner: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  nominatedConfigurer(overrides?: CallOverrides): Promise<string>;

  nominatedOwner(overrides?: CallOverrides): Promise<string>;

  owner(overrides?: CallOverrides): Promise<string>;

  renounceConfigurerNomination(
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  renounceNomination(
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  setConfigurer(
    newConfigurer: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  setCounter(
    _counter: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  callStatic: {
    acceptConfigurerRole(overrides?: CallOverrides): Promise<void>;

    acceptOwnership(overrides?: CallOverrides): Promise<void>;

    configurer(overrides?: CallOverrides): Promise<string>;

    countUp(overrides?: CallOverrides): Promise<void>;

    counter(overrides?: CallOverrides): Promise<BigNumber>;

    nominateNewConfigurer(
      newNominatedConfigurer: string,
      overrides?: CallOverrides
    ): Promise<void>;

    nominateNewOwner(
      newNominatedOwner: string,
      overrides?: CallOverrides
    ): Promise<void>;

    nominatedConfigurer(overrides?: CallOverrides): Promise<string>;

    nominatedOwner(overrides?: CallOverrides): Promise<string>;

    owner(overrides?: CallOverrides): Promise<string>;

    renounceConfigurerNomination(overrides?: CallOverrides): Promise<void>;

    renounceNomination(overrides?: CallOverrides): Promise<void>;

    setConfigurer(
      newConfigurer: string,
      overrides?: CallOverrides
    ): Promise<void>;

    setCounter(
      _counter: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "ConfigurerChanged(address,address)"(
      oldConfigurer?: null,
      newConfigurer?: null
    ): ConfigurerChangedEventFilter;
    ConfigurerChanged(
      oldConfigurer?: null,
      newConfigurer?: null
    ): ConfigurerChangedEventFilter;

    "ConfigurerNominated(address)"(
      newConfigurer?: null
    ): ConfigurerNominatedEventFilter;
    ConfigurerNominated(newConfigurer?: null): ConfigurerNominatedEventFilter;

    "OwnerChanged(address,address)"(
      oldOwner?: null,
      newOwner?: null
    ): OwnerChangedEventFilter;
    OwnerChanged(oldOwner?: null, newOwner?: null): OwnerChangedEventFilter;

    "OwnerNominated(address)"(newOwner?: null): OwnerNominatedEventFilter;
    OwnerNominated(newOwner?: null): OwnerNominatedEventFilter;
  };

  estimateGas: {
    acceptConfigurerRole(
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    acceptOwnership(
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    configurer(overrides?: CallOverrides): Promise<BigNumber>;

    countUp(overrides?: Overrides & { from?: string }): Promise<BigNumber>;

    counter(overrides?: CallOverrides): Promise<BigNumber>;

    nominateNewConfigurer(
      newNominatedConfigurer: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    nominateNewOwner(
      newNominatedOwner: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    nominatedConfigurer(overrides?: CallOverrides): Promise<BigNumber>;

    nominatedOwner(overrides?: CallOverrides): Promise<BigNumber>;

    owner(overrides?: CallOverrides): Promise<BigNumber>;

    renounceConfigurerNomination(
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    renounceNomination(
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    setConfigurer(
      newConfigurer: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    setCounter(
      _counter: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    acceptConfigurerRole(
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    acceptOwnership(
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    configurer(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    countUp(
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    counter(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    nominateNewConfigurer(
      newNominatedConfigurer: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    nominateNewOwner(
      newNominatedOwner: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    nominatedConfigurer(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    nominatedOwner(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    owner(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    renounceConfigurerNomination(
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    renounceNomination(
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    setConfigurer(
      newConfigurer: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    setCounter(
      _counter: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;
  };
}
