// @synthetixio/core-contracts/typechain-types/mocks/utils/ERC165HelperMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface ERC165HelperMockInterface extends utils.Interface {
  functions: {
    "getERC20InterfaceId()": FunctionFragment;
    "getERC721InterfaceId()": FunctionFragment;
    "supportsInterface(address,bytes4)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "getERC20InterfaceId"
      | "getERC721InterfaceId"
      | "supportsInterface"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "getERC20InterfaceId",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getERC721InterfaceId",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "supportsInterface",
    values: [string, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "getERC20InterfaceId",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getERC721InterfaceId",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "supportsInterface",
    data: BytesLike
  ): Result;

  events: {};
}

export interface ERC165HelperMock extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: ERC165HelperMockInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    getERC20InterfaceId(overrides?: CallOverrides): Promise<[string]>;

    getERC721InterfaceId(overrides?: CallOverrides): Promise<[string]>;

    supportsInterface(
      candidate: string,
      interfaceId: BytesLike,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;
  };

  getERC20InterfaceId(overrides?: CallOverrides): Promise<string>;

  getERC721InterfaceId(overrides?: CallOverrides): Promise<string>;

  supportsInterface(
    candidate: string,
    interfaceId: BytesLike,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  callStatic: {
    getERC20InterfaceId(overrides?: CallOverrides): Promise<string>;

    getERC721InterfaceId(overrides?: CallOverrides): Promise<string>;

    supportsInterface(
      candidate: string,
      interfaceId: BytesLike,
      overrides?: CallOverrides
    ): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    getERC20InterfaceId(overrides?: CallOverrides): Promise<BigNumber>;

    getERC721InterfaceId(overrides?: CallOverrides): Promise<BigNumber>;

    supportsInterface(
      candidate: string,
      interfaceId: BytesLike,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    getERC20InterfaceId(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getERC721InterfaceId(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    supportsInterface(
      candidate: string,
      interfaceId: BytesLike,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;
  };
}
