// @synthetixio/perps-market/typechain-types/factories/contracts/interfaces/IAccountEvents__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IAccountEvents,
  IAccountEventsInterface,
} from "../../../contracts/interfaces/IAccountEvents";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "amount",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "accountDebt",
        type: "uint256",
      },
    ],
    name: "AccountCharged",
    type: "event",
  },
] as const;

export class IAccountEvents__factory {
  static readonly abi = _abi;
  static createInterface(): IAccountEventsInterface {
    return new utils.Interface(_abi) as IAccountEventsInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IAccountEvents {
    return new Contract(address, _abi, signerOrProvider) as IAccountEvents;
  }
}
