// @synthetixio/perps-market/typechain-types/factories/contracts/interfaces/IAsyncOrderSettlementPythModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IAsyncOrderSettlementPythModule,
  IAsyncOrderSettlementPythModuleInterface,
} from "../../../contracts/interfaces/IAsyncOrderSettlementPythModule";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "interest",
        type: "uint256",
      },
    ],
    name: "InterestCharged",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: true,
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "fillPrice",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "pnl",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "accruedFunding",
        type: "int256",
      },
      {
        indexed: false,
        internalType: "int128",
        name: "sizeDelta",
        type: "int128",
      },
      {
        indexed: false,
        internalType: "int128",
        name: "newSize",
        type: "int128",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "totalFees",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "referralFees",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "collectedFees",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "settlementReward",
        type: "uint256",
      },
      {
        indexed: true,
        internalType: "bytes32",
        name: "trackingCode",
        type: "bytes32",
      },
      {
        indexed: false,
        internalType: "address",
        name: "settler",
        type: "address",
      },
    ],
    name: "OrderSettled",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "settleOrder",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IAsyncOrderSettlementPythModule__factory {
  static readonly abi = _abi;
  static createInterface(): IAsyncOrderSettlementPythModuleInterface {
    return new utils.Interface(
      _abi
    ) as IAsyncOrderSettlementPythModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IAsyncOrderSettlementPythModule {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IAsyncOrderSettlementPythModule;
  }
}
