// @synthetixio/perps-market/typechain-types/factories/contracts/interfaces/IDistributorErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IDistributorErrors,
  IDistributorErrorsInterface,
} from "../../../contracts/interfaces/IDistributorErrors";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "id",
        type: "uint128",
      },
      {
        internalType: "address",
        name: "distributor",
        type: "address",
      },
    ],
    name: "InvalidDistributor",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "distributor",
        type: "address",
      },
    ],
    name: "InvalidDistributorContract",
    type: "error",
  },
] as const;

export class IDistributorErrors__factory {
  static readonly abi = _abi;
  static createInterface(): IDistributorErrorsInterface {
    return new utils.Interface(_abi) as IDistributorErrorsInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IDistributorErrors {
    return new Contract(address, _abi, signerOrProvider) as IDistributorErrors;
  }
}
