// @synthetixio/perps-market/typechain-types/factories/contracts/interfaces/IPerpsMarketFactoryModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IPerpsMarketFactoryModule,
  IPerpsMarketFactoryModuleInterface,
} from "../../../contracts/interfaces/IPerpsMarketFactoryModule";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "uint128",
        name: "globalPerpsMarketId",
        type: "uint128",
      },
    ],
    name: "FactoryInitialized",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "perpsMarketId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "string",
        name: "marketName",
        type: "string",
      },
      {
        indexed: false,
        internalType: "string",
        name: "marketSymbol",
        type: "string",
      },
    ],
    name: "MarketCreated",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "requestedMarketId",
        type: "uint128",
      },
      {
        internalType: "string",
        name: "marketName",
        type: "string",
      },
      {
        internalType: "string",
        name: "marketSymbol",
        type: "string",
      },
    ],
    name: "createMarket",
    outputs: [
      {
        internalType: "uint128",
        name: "",
        type: "uint128",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "contract ISynthetixSystem",
        name: "synthetix",
        type: "address",
      },
      {
        internalType: "contract ISpotMarketSystem",
        name: "spotMarket",
        type: "address",
      },
    ],
    name: "initializeFactory",
    outputs: [
      {
        internalType: "uint128",
        name: "",
        type: "uint128",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "interestRate",
    outputs: [
      {
        internalType: "uint128",
        name: "rate",
        type: "uint128",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "minimumCredit",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "name",
    outputs: [
      {
        internalType: "string",
        name: "",
        type: "string",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "reportedDebt",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "string",
        name: "marketName",
        type: "string",
      },
    ],
    name: "setPerpsMarketName",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes4",
        name: "interfaceID",
        type: "bytes4",
      },
    ],
    name: "supportsInterface",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "utilizationRate",
    outputs: [
      {
        internalType: "uint256",
        name: "rate",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "delegatedCollateral",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "lockedCredit",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class IPerpsMarketFactoryModule__factory {
  static readonly abi = _abi;
  static createInterface(): IPerpsMarketFactoryModuleInterface {
    return new utils.Interface(_abi) as IPerpsMarketFactoryModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPerpsMarketFactoryModule {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as IPerpsMarketFactoryModule;
  }
}
