// @synthetixio/perps-market/typechain-types/factories/@synthetixio/core-contracts/contracts/token/ERC20Helper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ERC20Helper,
  ERC20HelperInterface,
} from "../../../../../@synthetixio/core-contracts/contracts/token/ERC20Helper";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "from",
        type: "address",
      },
      {
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "value",
        type: "uint256",
      },
    ],
    name: "FailedTransfer",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220f87f17fa074abe94814326b5eea66e2105f8300c5add8d547bc051d010952e0a64736f6c63430008160033";

type ERC20HelperConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ERC20HelperConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ERC20Helper__factory extends ContractFactory {
  constructor(...args: ERC20HelperConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ERC20Helper> {
    return super.deploy(overrides || {}) as Promise<ERC20Helper>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ERC20Helper {
    return super.attach(address) as ERC20Helper;
  }
  override connect(signer: Signer): ERC20Helper__factory {
    return super.connect(signer) as ERC20Helper__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ERC20HelperInterface {
    return new utils.Interface(_abi) as ERC20HelperInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ERC20Helper {
    return new Contract(address, _abi, signerOrProvider) as ERC20Helper;
  }
}
