// @synthetixio/core-modules/typechain-types/factories/contracts/interfaces/ISampleOwnedModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  ISampleOwnedModule,
  ISampleOwnedModuleInterface,
} from "../../../contracts/interfaces/ISampleOwnedModule";

const _abi = [
  {
    inputs: [],
    name: "getProtectedValue",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "newProtectedValue",
        type: "uint256",
      },
    ],
    name: "setProtectedValue",
    outputs: [],
    stateMutability: "payable",
    type: "function",
  },
] as const;

export class ISampleOwnedModule__factory {
  static readonly abi = _abi;
  static createInterface(): ISampleOwnedModuleInterface {
    return new utils.Interface(_abi) as ISampleOwnedModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ISampleOwnedModule {
    return new Contract(address, _abi, signerOrProvider) as ISampleOwnedModule;
  }
}
