// @synthetixio/core-modules/typechain-types/factories/contracts/interfaces/ISampleFeatureFlagModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  ISampleFeatureFlagModule,
  ISampleFeatureFlagModuleInterface,
} from "../../../contracts/interfaces/ISampleFeatureFlagModule";

const _abi = [
  {
    inputs: [],
    name: "getFeatureFlaggedValue",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "valueToSet",
        type: "uint256",
      },
    ],
    name: "setFeatureFlaggedValue",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ISampleFeatureFlagModule__factory {
  static readonly abi = _abi;
  static createInterface(): ISampleFeatureFlagModuleInterface {
    return new utils.Interface(_abi) as ISampleFeatureFlagModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ISampleFeatureFlagModule {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as ISampleFeatureFlagModule;
  }
}
