// @synthetixio/spot-market/typechain-types/contracts/interfaces/external/IPythERC7412Wrapper.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  PayableOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../../common";

export interface IPythERC7412WrapperInterface extends utils.Interface {
  functions: {
    "fulfillOracleQuery(bytes)": FunctionFragment;
    "getBenchmarkPrice(bytes32,uint64)": FunctionFragment;
    "getLatestPrice(bytes32,uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "fulfillOracleQuery"
      | "getBenchmarkPrice"
      | "getLatestPrice"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "fulfillOracleQuery",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getBenchmarkPrice",
    values: [BytesLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getLatestPrice",
    values: [BytesLike, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "fulfillOracleQuery",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getBenchmarkPrice",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getLatestPrice",
    data: BytesLike
  ): Result;

  events: {};
}

export interface IPythERC7412Wrapper extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IPythERC7412WrapperInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    fulfillOracleQuery(
      signedOffchainData: BytesLike,
      overrides?: PayableOverrides & { from?: string }
    ): Promise<ContractTransaction>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    getLatestPrice(
      priceId: BytesLike,
      stalenessTolerance: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;
  };

  fulfillOracleQuery(
    signedOffchainData: BytesLike,
    overrides?: PayableOverrides & { from?: string }
  ): Promise<ContractTransaction>;

  getBenchmarkPrice(
    priceId: BytesLike,
    requestedTime: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  getLatestPrice(
    priceId: BytesLike,
    stalenessTolerance: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  callStatic: {
    fulfillOracleQuery(
      signedOffchainData: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getLatestPrice(
      priceId: BytesLike,
      stalenessTolerance: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  filters: {};

  estimateGas: {
    fulfillOracleQuery(
      signedOffchainData: BytesLike,
      overrides?: PayableOverrides & { from?: string }
    ): Promise<BigNumber>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    getLatestPrice(
      priceId: BytesLike,
      stalenessTolerance: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    fulfillOracleQuery(
      signedOffchainData: BytesLike,
      overrides?: PayableOverrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getLatestPrice(
      priceId: BytesLike,
      stalenessTolerance: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
