// @synthetixio/core-modules/typechain-types/contracts/modules/mocks/GenericModule.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../../common";

export interface GenericModuleInterface extends utils.Interface {
  functions: {
    "getFortyTwo()": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "getFortyTwo"): FunctionFragment;

  encodeFunctionData(
    functionFragment: "getFortyTwo",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "getFortyTwo",
    data: BytesLike
  ): Result;

  events: {};
}

export interface GenericModule extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: GenericModuleInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    getFortyTwo(overrides?: CallOverrides): Promise<[BigNumber]>;
  };

  getFortyTwo(overrides?: CallOverrides): Promise<BigNumber>;

  callStatic: {
    getFortyTwo(overrides?: CallOverrides): Promise<BigNumber>;
  };

  filters: {};

  estimateGas: {
    getFortyTwo(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    getFortyTwo(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
