// @synthetixio/core-modules/typechain-types/contracts/modules/mocks/SampleOwnedModule.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  PayableOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../../common";

export interface SampleOwnedModuleInterface extends utils.Interface {
  functions: {
    "getProtectedValue()": FunctionFragment;
    "setProtectedValue(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "getProtectedValue" | "setProtectedValue"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "getProtectedValue",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setProtectedValue",
    values: [BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "getProtectedValue",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setProtectedValue",
    data: BytesLike
  ): Result;

  events: {};
}

export interface SampleOwnedModule extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: SampleOwnedModuleInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    getProtectedValue(overrides?: CallOverrides): Promise<[BigNumber]>;

    setProtectedValue(
      newProtectedValue: BigNumberish,
      overrides?: PayableOverrides & { from?: string }
    ): Promise<ContractTransaction>;
  };

  getProtectedValue(overrides?: CallOverrides): Promise<BigNumber>;

  setProtectedValue(
    newProtectedValue: BigNumberish,
    overrides?: PayableOverrides & { from?: string }
  ): Promise<ContractTransaction>;

  callStatic: {
    getProtectedValue(overrides?: CallOverrides): Promise<BigNumber>;

    setProtectedValue(
      newProtectedValue: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    getProtectedValue(overrides?: CallOverrides): Promise<BigNumber>;

    setProtectedValue(
      newProtectedValue: BigNumberish,
      overrides?: PayableOverrides & { from?: string }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    getProtectedValue(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    setProtectedValue(
      newProtectedValue: BigNumberish,
      overrides?: PayableOverrides & { from?: string }
    ): Promise<PopulatedTransaction>;
  };
}
