// @synthetixio/core-modules/typechain-types/factories/@synthetixio/core-contracts/contracts/errors/AccessError__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  AccessError,
  AccessErrorInterface,
} from "../../../../../@synthetixio/core-contracts/contracts/errors/AccessError";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "addr",
        type: "address",
      },
    ],
    name: "Unauthorized",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212209c22e282cf2887515849a85a18eeef10620ad1af4a3f297dec1d058a717d055764736f6c63430008160033";

type AccessErrorConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AccessErrorConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AccessError__factory extends ContractFactory {
  constructor(...args: AccessErrorConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<AccessError> {
    return super.deploy(overrides || {}) as Promise<AccessError>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): AccessError {
    return super.attach(address) as AccessError;
  }
  override connect(signer: Signer): AccessError__factory {
    return super.connect(signer) as AccessError__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AccessErrorInterface {
    return new utils.Interface(_abi) as AccessErrorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AccessError {
    return new Contract(address, _abi, signerOrProvider) as AccessError;
  }
}
