// @synthetixio/core-modules/typechain-types/factories/@synthetixio/core-contracts/contracts/errors/InitError__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  InitError,
  InitErrorInterface,
} from "../../../../../@synthetixio/core-contracts/contracts/errors/InitError";

const _abi = [
  {
    inputs: [],
    name: "AlreadyInitialized",
    type: "error",
  },
  {
    inputs: [],
    name: "NotInitialized",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220b0cf3c960116e0826e23a3947c6e8f379a9a7c919328109d47364a6d7298c63c64736f6c63430008160033";

type InitErrorConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: InitErrorConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class InitError__factory extends ContractFactory {
  constructor(...args: InitErrorConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<InitError> {
    return super.deploy(overrides || {}) as Promise<InitError>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): InitError {
    return super.attach(address) as InitError;
  }
  override connect(signer: Signer): InitError__factory {
    return super.connect(signer) as InitError__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): InitErrorInterface {
    return new utils.Interface(_abi) as InitErrorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): InitError {
    return new Contract(address, _abi, signerOrProvider) as InitError;
  }
}
