// @synthetixio/core-modules/typechain-types/factories/@synthetixio/core-contracts/contracts/errors/ParameterError__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ParameterError,
  ParameterErrorInterface,
} from "../../../../../@synthetixio/core-contracts/contracts/errors/ParameterError";

const _abi = [
  {
    inputs: [
      {
        internalType: "string",
        name: "parameter",
        type: "string",
      },
      {
        internalType: "string",
        name: "reason",
        type: "string",
      },
    ],
    name: "InvalidParameter",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220b531d61036026e7536534f7155d3a80eb41afae52c9bcf08082c300006c503bd64736f6c63430008160033";

type ParameterErrorConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ParameterErrorConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ParameterError__factory extends ContractFactory {
  constructor(...args: ParameterErrorConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ParameterError> {
    return super.deploy(overrides || {}) as Promise<ParameterError>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ParameterError {
    return super.attach(address) as ParameterError;
  }
  override connect(signer: Signer): ParameterError__factory {
    return super.connect(signer) as ParameterError__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ParameterErrorInterface {
    return new utils.Interface(_abi) as ParameterErrorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ParameterError {
    return new Contract(address, _abi, signerOrProvider) as ParameterError;
  }
}
