// @synthetixio/spot-market/typechain-types/factories/@synthetixio/oracle-manager/contracts/storage/NodeDefinition__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  NodeDefinition,
  NodeDefinitionInterface,
} from "../../../../../@synthetixio/oracle-manager/contracts/storage/NodeDefinition";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "nodeId",
        type: "bytes32",
      },
    ],
    name: "UnprocessableNode",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220b4bd869ed509e79df18970afc9f7841729b1c1e5770208df288fd4244f16692064736f6c63430008160033";

type NodeDefinitionConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: NodeDefinitionConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class NodeDefinition__factory extends ContractFactory {
  constructor(...args: NodeDefinitionConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<NodeDefinition> {
    return super.deploy(overrides || {}) as Promise<NodeDefinition>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): NodeDefinition {
    return super.attach(address) as NodeDefinition;
  }
  override connect(signer: Signer): NodeDefinition__factory {
    return super.connect(signer) as NodeDefinition__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): NodeDefinitionInterface {
    return new utils.Interface(_abi) as NodeDefinitionInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): NodeDefinition {
    return new Contract(address, _abi, signerOrProvider) as NodeDefinition;
  }
}
