// @synthetixio/spot-market/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastU32__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastU32,
  SafeCastU32Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastU32";

const _abi = [
  {
    inputs: [],
    name: "OverflowUint32ToInt32",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220d84b9f7c46e0bbd2a9ee668b98ec0a33df1b84b7d16e33b330bf276b060b5d2364736f6c63430008160033";

type SafeCastU32ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastU32ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastU32__factory extends ContractFactory {
  constructor(...args: SafeCastU32ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastU32> {
    return super.deploy(overrides || {}) as Promise<SafeCastU32>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastU32 {
    return super.attach(address) as SafeCastU32;
  }
  override connect(signer: Signer): SafeCastU32__factory {
    return super.connect(signer) as SafeCastU32__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastU32Interface {
    return new utils.Interface(_abi) as SafeCastU32Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastU32 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastU32;
  }
}
