// @synthetixio/spot-market/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI128__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastI128,
  SafeCastI128Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI128";

const _abi = [
  {
    inputs: [],
    name: "OverflowInt128ToInt32",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowInt128ToUint128",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220f9e3231478891daa760405b592206714ad437f641f9942ca4526f30371ce378b64736f6c63430008160033";

type SafeCastI128ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastI128ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastI128__factory extends ContractFactory {
  constructor(...args: SafeCastI128ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastI128> {
    return super.deploy(overrides || {}) as Promise<SafeCastI128>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastI128 {
    return super.attach(address) as SafeCastI128;
  }
  override connect(signer: Signer): SafeCastI128__factory {
    return super.connect(signer) as SafeCastI128__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastI128Interface {
    return new utils.Interface(_abi) as SafeCastI128Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastI128 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastI128;
  }
}
