// @synthetixio/perps-market/typechain-types/factories/@synthetixio/spot-market/contracts/utils/TransactionUtil.sol/Transaction__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  Transaction,
  TransactionInterface,
} from "../../../../../../@synthetixio/spot-market/contracts/utils/TransactionUtil.sol/Transaction";

const _abi = [
  {
    inputs: [
      {
        internalType: "enum Transaction.Type",
        name: "transactionType",
        type: "uint8",
      },
    ],
    name: "InvalidAsyncTransactionType",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "txnType",
        type: "uint128",
      },
    ],
    name: "InvalidTransactionTypeIndex",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212207424e6b6dd6bc5a9ea7c04cc7e7d33188970310b9b56686182250f5494e378d764736f6c63430008160033";

type TransactionConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: TransactionConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class Transaction__factory extends ContractFactory {
  constructor(...args: TransactionConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<Transaction> {
    return super.deploy(overrides || {}) as Promise<Transaction>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): Transaction {
    return super.attach(address) as Transaction;
  }
  override connect(signer: Signer): Transaction__factory {
    return super.connect(signer) as Transaction__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): TransactionInterface {
    return new utils.Interface(_abi) as TransactionInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Transaction {
    return new Contract(address, _abi, signerOrProvider) as Transaction;
  }
}
