// @lerna-lite/core/dist/utils/collect-updates/lib/make-diff-predicate.js
import { globbySync } from 'globby';
import log from 'npmlog';
import { filter as minimatchFilter } from 'minimatch';
import { dirname, relative } from 'node:path';
import slash from 'slash';
import { execSync } from '../../../child-process.js';
export function makeDiffPredicate(committish, execOpts, ignorePatterns = [], diffOpts) {
    const ignoreFilters = new Set(ignorePatterns.map((p) => minimatchFilter(`!${p}`, {
        matchBase: true,
        dot: true,
    })));
    if (ignoreFilters.size) {
        log.info('ignoring diff in paths matching', ignorePatterns.join(' '));
    }
    return function hasDiffSinceThatIsntIgnored(node) {
        const diff = diffSinceIn(committish, node.location, execOpts, diffOpts);
        if (diff === '') {
            log.silly('', 'no diff found in %s', node.name);
            return false;
        }
        log.silly('found diff in', diff);
        let changedFiles = diff.split('\n');
        if (ignoreFilters.size) {
            for (const ignored of ignoreFilters) {
                changedFiles = changedFiles.filter(ignored);
            }
        }
        if (changedFiles.length) {
            log.verbose('filtered diff', changedFiles.join(' '));
        }
        else {
            log.verbose('', 'no diff found in %s (after filtering)', node.name);
        }
        return changedFiles.length > 0;
    };
}
function diffSinceIn(committish, location, execOpts, diffOpts) {
    const args = ['diff', '--name-only', committish];
    const formattedLocation = slash(relative(execOpts.cwd, location));
    if (formattedLocation) {
        let independentSubpackages = [];
        if (diffOpts?.independentSubpackages) {
            independentSubpackages = globbySync('**/*/package.json', {
                cwd: formattedLocation,
                nodir: true,
                ignore: ['**/node_modules/**'],
            }).map((file) => `:^${formattedLocation}/${dirname(file)}`);
        }
        args.push('--', formattedLocation, ...independentSubpackages);
    }
    log.silly('checking diff', formattedLocation);
    return execSync('git', args, execOpts);
}
