// @synthetixio/spot-market/typechain-types/factories/contracts/interfaces/IAsyncOrderModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IAsyncOrderModule,
  IAsyncOrderModuleInterface,
} from "../../../contracts/interfaces/IAsyncOrderModule";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: true,
        internalType: "uint128",
        name: "asyncOrderId",
        type: "uint128",
      },
      {
        components: [
          {
            internalType: "uint128",
            name: "id",
            type: "uint128",
          },
          {
            internalType: "address",
            name: "owner",
            type: "address",
          },
          {
            internalType: "enum Transaction.Type",
            name: "orderType",
            type: "uint8",
          },
          {
            internalType: "uint256",
            name: "amountEscrowed",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "settlementStrategyId",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "commitmentTime",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "minimumSettlementAmount",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "settledAt",
            type: "uint256",
          },
          {
            internalType: "address",
            name: "referrer",
            type: "address",
          },
        ],
        indexed: false,
        internalType: "struct AsyncOrderClaim.Data",
        name: "asyncOrderClaim",
        type: "tuple",
      },
      {
        indexed: true,
        internalType: "address",
        name: "sender",
        type: "address",
      },
    ],
    name: "OrderCancelled",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: true,
        internalType: "enum Transaction.Type",
        name: "orderType",
        type: "uint8",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amountProvided",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint128",
        name: "asyncOrderId",
        type: "uint128",
      },
      {
        indexed: true,
        internalType: "address",
        name: "sender",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "referrer",
        type: "address",
      },
    ],
    name: "OrderCommitted",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "asyncOrderId",
        type: "uint128",
      },
    ],
    name: "cancelOrder",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "enum Transaction.Type",
        name: "orderType",
        type: "uint8",
      },
      {
        internalType: "uint256",
        name: "amountProvided",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "settlementStrategyId",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "minimumSettlementAmount",
        type: "uint256",
      },
      {
        internalType: "address",
        name: "referrer",
        type: "address",
      },
    ],
    name: "commitOrder",
    outputs: [
      {
        components: [
          {
            internalType: "uint128",
            name: "id",
            type: "uint128",
          },
          {
            internalType: "address",
            name: "owner",
            type: "address",
          },
          {
            internalType: "enum Transaction.Type",
            name: "orderType",
            type: "uint8",
          },
          {
            internalType: "uint256",
            name: "amountEscrowed",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "settlementStrategyId",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "commitmentTime",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "minimumSettlementAmount",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "settledAt",
            type: "uint256",
          },
          {
            internalType: "address",
            name: "referrer",
            type: "address",
          },
        ],
        internalType: "struct AsyncOrderClaim.Data",
        name: "asyncOrderClaim",
        type: "tuple",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "asyncOrderId",
        type: "uint128",
      },
    ],
    name: "getAsyncOrderClaim",
    outputs: [
      {
        components: [
          {
            internalType: "uint128",
            name: "id",
            type: "uint128",
          },
          {
            internalType: "address",
            name: "owner",
            type: "address",
          },
          {
            internalType: "enum Transaction.Type",
            name: "orderType",
            type: "uint8",
          },
          {
            internalType: "uint256",
            name: "amountEscrowed",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "settlementStrategyId",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "commitmentTime",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "minimumSettlementAmount",
            type: "uint256",
          },
          {
            internalType: "uint256",
            name: "settledAt",
            type: "uint256",
          },
          {
            internalType: "address",
            name: "referrer",
            type: "address",
          },
        ],
        internalType: "struct AsyncOrderClaim.Data",
        name: "asyncOrderClaim",
        type: "tuple",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class IAsyncOrderModule__factory {
  static readonly abi = _abi;
  static createInterface(): IAsyncOrderModuleInterface {
    return new utils.Interface(_abi) as IAsyncOrderModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IAsyncOrderModule {
    return new Contract(address, _abi, signerOrProvider) as IAsyncOrderModule;
  }
}
