// @synthetixio/spot-market/typechain-types/factories/contracts/interfaces/ISpotMarketFactoryModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  ISpotMarketFactoryModule,
  ISpotMarketFactoryModuleInterface,
} from "../../../contracts/interfaces/ISpotMarketFactoryModule";

const _abi = [
  {
    inputs: [],
    name: "InvalidMarketOwner",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "addr",
        type: "address",
      },
    ],
    name: "NotNominated",
    type: "error",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "rate",
        type: "uint256",
      },
    ],
    name: "DecayRateUpdated",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "address",
        name: "nominee",
        type: "address",
      },
    ],
    name: "MarketNominationRenounced",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "address",
        name: "oldOwner",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "MarketOwnerChanged",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "MarketOwnerNominated",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "address",
        name: "synthImplementation",
        type: "address",
      },
    ],
    name: "SynthImplementationSet",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint256",
        name: "synthMarketId",
        type: "uint256",
      },
      {
        indexed: true,
        internalType: "address",
        name: "proxy",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "implementation",
        type: "address",
      },
    ],
    name: "SynthImplementationUpgraded",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint256",
        name: "synthMarketId",
        type: "uint256",
      },
      {
        indexed: true,
        internalType: "bytes32",
        name: "buyFeedId",
        type: "bytes32",
      },
      {
        indexed: true,
        internalType: "bytes32",
        name: "sellFeedId",
        type: "bytes32",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "strictStalenessTolerance",
        type: "uint256",
      },
    ],
    name: "SynthPriceDataUpdated",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint256",
        name: "synthMarketId",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "address",
        name: "synthTokenAddress",
        type: "address",
      },
    ],
    name: "SynthRegistered",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "address",
        name: "synthetix",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "usdTokenAddress",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "oracleManager",
        type: "address",
      },
    ],
    name: "SynthetixSystemSet",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
    ],
    name: "acceptMarketOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "string",
        name: "tokenName",
        type: "string",
      },
      {
        internalType: "string",
        name: "tokenSymbol",
        type: "string",
      },
      {
        internalType: "address",
        name: "synthOwner",
        type: "address",
      },
    ],
    name: "createSynth",
    outputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
    ],
    name: "getMarketOwner",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
    ],
    name: "getNominatedMarketOwner",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "getPriceData",
    outputs: [
      {
        internalType: "bytes32",
        name: "buyFeedId",
        type: "bytes32",
      },
      {
        internalType: "bytes32",
        name: "sellFeedId",
        type: "bytes32",
      },
      {
        internalType: "uint256",
        name: "strictPriceStalenessTolerance",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "getSynth",
    outputs: [
      {
        internalType: "address",
        name: "synthAddress",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "getSynthImpl",
    outputs: [
      {
        internalType: "address",
        name: "implAddress",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "transactionType",
        type: "uint128",
      },
      {
        internalType: "enum Price.Tolerance",
        name: "priceTolerance",
        type: "uint8",
      },
    ],
    name: "indexPrice",
    outputs: [
      {
        internalType: "uint256",
        name: "price",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "minimumCredit",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "name",
    outputs: [
      {
        internalType: "string",
        name: "",
        type: "string",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
      {
        internalType: "address",
        name: "newNominatedOwner",
        type: "address",
      },
    ],
    name: "nominateMarketOwner",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
    ],
    name: "renounceMarketNomination",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "synthMarketId",
        type: "uint128",
      },
    ],
    name: "renounceMarketOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "reportedDebt",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "rate",
        type: "uint256",
      },
    ],
    name: "setDecayRate",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "synthImplementation",
        type: "address",
      },
    ],
    name: "setSynthImplementation",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "contract ISynthetixSystem",
        name: "synthetix",
        type: "address",
      },
    ],
    name: "setSynthetix",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes4",
        name: "interfaceID",
        type: "bytes4",
      },
    ],
    name: "supportsInterface",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
      {
        internalType: "bytes32",
        name: "buyFeedId",
        type: "bytes32",
      },
      {
        internalType: "bytes32",
        name: "sellFeedId",
        type: "bytes32",
      },
      {
        internalType: "uint256",
        name: "strictPriceStalenessTolerance",
        type: "uint256",
      },
    ],
    name: "updatePriceData",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "upgradeSynthImpl",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class ISpotMarketFactoryModule__factory {
  static readonly abi = _abi;
  static createInterface(): ISpotMarketFactoryModuleInterface {
    return new utils.Interface(_abi) as ISpotMarketFactoryModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ISpotMarketFactoryModule {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as ISpotMarketFactoryModule;
  }
}
