// @synthetixio/spot-market/typechain-types/factories/contracts/storage/AsyncOrder__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  AsyncOrder,
  AsyncOrderInterface,
} from "../../../contracts/storage/AsyncOrder";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "expected",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "actual",
        type: "uint256",
      },
    ],
    name: "InsufficientSharesAmount",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220c1fb2a595e53df871c1f984260dbbd9e1736e8bd3ddd0c0a9242d89af8362a3164736f6c63430008160033";

type AsyncOrderConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AsyncOrderConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AsyncOrder__factory extends ContractFactory {
  constructor(...args: AsyncOrderConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<AsyncOrder> {
    return super.deploy(overrides || {}) as Promise<AsyncOrder>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): AsyncOrder {
    return super.attach(address) as AsyncOrder;
  }
  override connect(signer: Signer): AsyncOrder__factory {
    return super.connect(signer) as AsyncOrder__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AsyncOrderInterface {
    return new utils.Interface(_abi) as AsyncOrderInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AsyncOrder {
    return new Contract(address, _abi, signerOrProvider) as AsyncOrder;
  }
}
