// @synthetixio/spot-market/typechain-types/factories/contracts/storage/AsyncOrderConfiguration__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  AsyncOrderConfiguration,
  AsyncOrderConfigurationInterface,
} from "../../../contracts/storage/AsyncOrderConfiguration";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "settlementStrategyId",
        type: "uint256",
      },
    ],
    name: "InvalidSettlementStrategy",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212202b662f2b45bb113c2a8d2409dbf73b1c38e5a698dea0639ff1b5531ae369a2d464736f6c63430008160033";

type AsyncOrderConfigurationConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AsyncOrderConfigurationConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AsyncOrderConfiguration__factory extends ContractFactory {
  constructor(...args: AsyncOrderConfigurationConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<AsyncOrderConfiguration> {
    return super.deploy(overrides || {}) as Promise<AsyncOrderConfiguration>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): AsyncOrderConfiguration {
    return super.attach(address) as AsyncOrderConfiguration;
  }
  override connect(signer: Signer): AsyncOrderConfiguration__factory {
    return super.connect(signer) as AsyncOrderConfiguration__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AsyncOrderConfigurationInterface {
    return new utils.Interface(_abi) as AsyncOrderConfigurationInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AsyncOrderConfiguration {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as AsyncOrderConfiguration;
  }
}
