// @synthetixio/spot-market/typechain-types/factories/contracts/storage/MarketConfiguration__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  MarketConfiguration,
  MarketConfigurationInterface,
} from "../../../contracts/storage/MarketConfiguration";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    name: "InvalidCollateralLeverage",
    type: "error",
  },
  {
    inputs: [],
    name: "InvalidUtilizationLeverage",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122095cbd3317bd9f174479f5e4680613ad9d32a1d11d817e41e3f157f5d8ad4d75964736f6c63430008160033";

type MarketConfigurationConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: MarketConfigurationConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class MarketConfiguration__factory extends ContractFactory {
  constructor(...args: MarketConfigurationConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<MarketConfiguration> {
    return super.deploy(overrides || {}) as Promise<MarketConfiguration>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): MarketConfiguration {
    return super.attach(address) as MarketConfiguration;
  }
  override connect(signer: Signer): MarketConfiguration__factory {
    return super.connect(signer) as MarketConfiguration__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): MarketConfigurationInterface {
    return new utils.Interface(_abi) as MarketConfigurationInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): MarketConfiguration {
    return new Contract(address, _abi, signerOrProvider) as MarketConfiguration;
  }
}
