// @synthetixio/spot-market/typechain-types/factories/contracts/storage/SettlementStrategy__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SettlementStrategy,
  SettlementStrategyInterface,
} from "../../../contracts/storage/SettlementStrategy";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "minimumAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "InvalidCommitmentAmount",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220408385108887e1d8dd69b0513c33417a6d8c7e190a8b9f360dbb734595fd8ced64736f6c63430008160033";

type SettlementStrategyConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SettlementStrategyConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SettlementStrategy__factory extends ContractFactory {
  constructor(...args: SettlementStrategyConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SettlementStrategy> {
    return super.deploy(overrides || {}) as Promise<SettlementStrategy>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SettlementStrategy {
    return super.attach(address) as SettlementStrategy;
  }
  override connect(signer: Signer): SettlementStrategy__factory {
    return super.connect(signer) as SettlementStrategy__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SettlementStrategyInterface {
    return new utils.Interface(_abi) as SettlementStrategyInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SettlementStrategy {
    return new Contract(address, _abi, signerOrProvider) as SettlementStrategy;
  }
}
