// @synthetixio/spot-market/typechain-types/factories/contracts/storage/SpotMarketFactory__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SpotMarketFactory,
  SpotMarketFactoryInterface,
} from "../../../contracts/storage/SpotMarketFactory";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "InvalidMarket",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "synthImplementation",
        type: "uint256",
      },
    ],
    name: "InvalidSynthImplementation",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "marketOwner",
        type: "address",
      },
      {
        internalType: "address",
        name: "sender",
        type: "address",
      },
    ],
    name: "OnlyMarketOwner",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212201df8cf6f6c2e2ac0e6c2e26b620ea6e970c4c5a689caeab96acfdad04de0bdee64736f6c63430008160033";

type SpotMarketFactoryConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SpotMarketFactoryConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SpotMarketFactory__factory extends ContractFactory {
  constructor(...args: SpotMarketFactoryConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SpotMarketFactory> {
    return super.deploy(overrides || {}) as Promise<SpotMarketFactory>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SpotMarketFactory {
    return super.attach(address) as SpotMarketFactory;
  }
  override connect(signer: Signer): SpotMarketFactory__factory {
    return super.connect(signer) as SpotMarketFactory__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SpotMarketFactoryInterface {
    return new utils.Interface(_abi) as SpotMarketFactoryInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SpotMarketFactory {
    return new Contract(address, _abi, signerOrProvider) as SpotMarketFactory;
  }
}
