// @synthetixio/spot-market/typechain-types/factories/contracts/storage/Wrapper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  Wrapper,
  WrapperInterface,
} from "../../../contracts/storage/Wrapper";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "configuredCollateralType",
        type: "address",
      },
    ],
    name: "InvalidCollateralType",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "maxWrappableAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "currentSupply",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountToWrap",
        type: "uint256",
      },
    ],
    name: "WrapperExceedsMaxAmount",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122016b1ef3c64665f13a6dbb70ce21b86b4ce7e55810cde3ce1a19988d0c3a79df364736f6c63430008160033";

type WrapperConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: WrapperConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class Wrapper__factory extends ContractFactory {
  constructor(...args: WrapperConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(overrides?: Overrides & { from?: string }): Promise<Wrapper> {
    return super.deploy(overrides || {}) as Promise<Wrapper>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): Wrapper {
    return super.attach(address) as Wrapper;
  }
  override connect(signer: Signer): Wrapper__factory {
    return super.connect(signer) as Wrapper__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): WrapperInterface {
    return new utils.Interface(_abi) as WrapperInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Wrapper {
    return new Contract(address, _abi, signerOrProvider) as Wrapper;
  }
}
