// @synthetixio/core-contracts/typechain-types/mocks/initializable/InitializableMock.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface InitializableMockInterface extends utils.Interface {
  functions: {
    "doubleValue()": FunctionFragment;
    "getNonCriticalValue()": FunctionFragment;
    "getValue()": FunctionFragment;
    "initializeInitializableMock(uint256)": FunctionFragment;
    "isInitializableMockInitialized()": FunctionFragment;
    "setNonCriticalValue(uint256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "doubleValue"
      | "getNonCriticalValue"
      | "getValue"
      | "initializeInitializableMock"
      | "isInitializableMockInitialized"
      | "setNonCriticalValue"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "doubleValue",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getNonCriticalValue",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "getValue", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "initializeInitializableMock",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "isInitializableMockInitialized",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setNonCriticalValue",
    values: [BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "doubleValue",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getNonCriticalValue",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "getValue", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "initializeInitializableMock",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isInitializableMockInitialized",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setNonCriticalValue",
    data: BytesLike
  ): Result;

  events: {};
}

export interface InitializableMock extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: InitializableMockInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    doubleValue(
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    getNonCriticalValue(overrides?: CallOverrides): Promise<[BigNumber]>;

    getValue(overrides?: CallOverrides): Promise<[BigNumber]>;

    initializeInitializableMock(
      initialValue: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    isInitializableMockInitialized(
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    setNonCriticalValue(
      nonCriticalValue: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;
  };

  doubleValue(
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  getNonCriticalValue(overrides?: CallOverrides): Promise<BigNumber>;

  getValue(overrides?: CallOverrides): Promise<BigNumber>;

  initializeInitializableMock(
    initialValue: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  isInitializableMockInitialized(overrides?: CallOverrides): Promise<boolean>;

  setNonCriticalValue(
    nonCriticalValue: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  callStatic: {
    doubleValue(overrides?: CallOverrides): Promise<void>;

    getNonCriticalValue(overrides?: CallOverrides): Promise<BigNumber>;

    getValue(overrides?: CallOverrides): Promise<BigNumber>;

    initializeInitializableMock(
      initialValue: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    isInitializableMockInitialized(overrides?: CallOverrides): Promise<boolean>;

    setNonCriticalValue(
      nonCriticalValue: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    doubleValue(overrides?: Overrides & { from?: string }): Promise<BigNumber>;

    getNonCriticalValue(overrides?: CallOverrides): Promise<BigNumber>;

    getValue(overrides?: CallOverrides): Promise<BigNumber>;

    initializeInitializableMock(
      initialValue: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    isInitializableMockInitialized(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    setNonCriticalValue(
      nonCriticalValue: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    doubleValue(
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    getNonCriticalValue(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    getValue(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    initializeInitializableMock(
      initialValue: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    isInitializableMockInitialized(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    setNonCriticalValue(
      nonCriticalValue: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;
  };
}
