// hardhat-cannon/dist/internal/augment-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.augmentProvider = void 0;
const builder_1 = require("@usecannon/builder");
function augmentProvider(originalProvider, outputs) {
    const provider = originalProvider.extend((0, builder_1.traceActions)(outputs));
    // Monkey patch to call original cannon extended estimateGas fn
    const originalRequest = provider.request;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    provider.request = async function request(args) {
        if (args.method === 'eth_estimateGas') {
            return await provider.estimateGas({
                maxFeePerGas: args.params[0].maxFeePerGas,
                maxPriorityFeePerGas: args.params[0].maxPriorityFeePerGas,
                account: args.params[0].from,
                to: args.params[0].to,
                data: args.params[0].data,
                value: args.params[0].value,
            });
        }
        return await originalRequest(args);
    };
    return provider;
}
exports.augmentProvider = augmentProvider;
