// @synthetixio/perps-market/typechain-types/factories/contracts/storage/GlobalPerpsMarket__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  GlobalPerpsMarket,
  GlobalPerpsMarketInterface,
} from "../../../contracts/storage/GlobalPerpsMarket";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "collateralAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "withdrawAmount",
        type: "uint256",
      },
    ],
    name: "InsufficientCollateral",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "maxAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "collateralAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "depositAmount",
        type: "uint256",
      },
    ],
    name: "MaxCollateralExceeded",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
    ],
    name: "SynthNotEnabledForCollateral",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220c0e3e03e66715299238acba936b8c45fd686396034a65c6cde80d15c58174d7064736f6c63430008160033";

type GlobalPerpsMarketConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: GlobalPerpsMarketConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class GlobalPerpsMarket__factory extends ContractFactory {
  constructor(...args: GlobalPerpsMarketConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<GlobalPerpsMarket> {
    return super.deploy(overrides || {}) as Promise<GlobalPerpsMarket>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): GlobalPerpsMarket {
    return super.attach(address) as GlobalPerpsMarket;
  }
  override connect(signer: Signer): GlobalPerpsMarket__factory {
    return super.connect(signer) as GlobalPerpsMarket__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): GlobalPerpsMarketInterface {
    return new utils.Interface(_abi) as GlobalPerpsMarketInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): GlobalPerpsMarket {
    return new Contract(address, _abi, signerOrProvider) as GlobalPerpsMarket;
  }
}
