// @synthetixio/perps-market/typechain-types/factories/contracts/storage/PerpsCollateralConfiguration__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  PerpsCollateralConfiguration,
  PerpsCollateralConfigurationInterface,
} from "../../../contracts/storage/PerpsCollateralConfiguration";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "id",
        type: "uint128",
      },
    ],
    name: "InvalidId",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220fd2619da9ee6f1bbcf199d52ba16fef20eb11175bcb57617d1b25fe6153fb56f64736f6c63430008160033";

type PerpsCollateralConfigurationConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: PerpsCollateralConfigurationConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class PerpsCollateralConfiguration__factory extends ContractFactory {
  constructor(...args: PerpsCollateralConfigurationConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<PerpsCollateralConfiguration> {
    return super.deploy(
      overrides || {}
    ) as Promise<PerpsCollateralConfiguration>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): PerpsCollateralConfiguration {
    return super.attach(address) as PerpsCollateralConfiguration;
  }
  override connect(signer: Signer): PerpsCollateralConfiguration__factory {
    return super.connect(signer) as PerpsCollateralConfiguration__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): PerpsCollateralConfigurationInterface {
    return new utils.Interface(_abi) as PerpsCollateralConfigurationInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): PerpsCollateralConfiguration {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as PerpsCollateralConfiguration;
  }
}
