// @synthetixio/perps-market/typechain-types/factories/contracts/storage/PerpsMarket__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  PerpsMarket,
  PerpsMarketInterface,
} from "../../../contracts/storage/PerpsMarket";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "InvalidMarket",
    type: "error",
  },
  {
    inputs: [],
    name: "KeeperCostsNotSet",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "PriceFeedNotSet",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220a4c13e3a64c13110f96b957193798f70b9721ad6b7df7abccc1e31bc3202c61e64736f6c63430008160033";

type PerpsMarketConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: PerpsMarketConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class PerpsMarket__factory extends ContractFactory {
  constructor(...args: PerpsMarketConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<PerpsMarket> {
    return super.deploy(overrides || {}) as Promise<PerpsMarket>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): PerpsMarket {
    return super.attach(address) as PerpsMarket;
  }
  override connect(signer: Signer): PerpsMarket__factory {
    return super.connect(signer) as PerpsMarket__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): PerpsMarketInterface {
    return new utils.Interface(_abi) as PerpsMarketInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): PerpsMarket {
    return new Contract(address, _abi, signerOrProvider) as PerpsMarket;
  }
}
