// @synthetixio/perps-market/typechain-types/factories/contracts/storage/PerpsMarketFactory__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  PerpsMarketFactory,
  PerpsMarketFactoryInterface,
} from "../../../contracts/storage/PerpsMarketFactory";

const _abi = [
  {
    inputs: [],
    name: "PerpsMarketAlreadyInitialized",
    type: "error",
  },
  {
    inputs: [],
    name: "PerpsMarketNotInitialized",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212206f98ab1daa41381b5d653701deab0b83fefe9dacf014cf0988fe633e8549fc2264736f6c63430008160033";

type PerpsMarketFactoryConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: PerpsMarketFactoryConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class PerpsMarketFactory__factory extends ContractFactory {
  constructor(...args: PerpsMarketFactoryConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<PerpsMarketFactory> {
    return super.deploy(overrides || {}) as Promise<PerpsMarketFactory>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): PerpsMarketFactory {
    return super.attach(address) as PerpsMarketFactory;
  }
  override connect(signer: Signer): PerpsMarketFactory__factory {
    return super.connect(signer) as PerpsMarketFactory__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): PerpsMarketFactoryInterface {
    return new utils.Interface(_abi) as PerpsMarketFactoryInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): PerpsMarketFactory {
    return new Contract(address, _abi, signerOrProvider) as PerpsMarketFactory;
  }
}
