// @synthetixio/perps-market/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastU56__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastU56,
  SafeCastU56Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastU56";

const _abi = [
  {
    inputs: [],
    name: "OverflowUint56ToInt56",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212201b42160831dfd6360eb2e99adc6956cebfd338ba2bc8b3569305c38032c439c464736f6c63430008160033";

type SafeCastU56ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastU56ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastU56__factory extends ContractFactory {
  constructor(...args: SafeCastU56ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastU56> {
    return super.deploy(overrides || {}) as Promise<SafeCastU56>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastU56 {
    return super.attach(address) as SafeCastU56;
  }
  override connect(signer: Signer): SafeCastU56__factory {
    return super.connect(signer) as SafeCastU56__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastU56Interface {
    return new utils.Interface(_abi) as SafeCastU56Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastU56 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastU56;
  }
}
