// @synthetixio/perps-market/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI256__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastI256,
  SafeCastI256Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI256";

const _abi = [
  {
    inputs: [],
    name: "OverflowInt256ToInt128",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowInt256ToInt24",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowInt256ToUint256",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220e9b5d541b35130783452c3cb53089ba3dd556560abb4f474291341f18d531b1964736f6c63430008160033";

type SafeCastI256ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastI256ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastI256__factory extends ContractFactory {
  constructor(...args: SafeCastI256ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastI256> {
    return super.deploy(overrides || {}) as Promise<SafeCastI256>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastI256 {
    return super.attach(address) as SafeCastI256;
  }
  override connect(signer: Signer): SafeCastI256__factory {
    return super.connect(signer) as SafeCastI256__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastI256Interface {
    return new utils.Interface(_abi) as SafeCastI256Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastI256 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastI256;
  }
}
