// @synthetixio/perps-market/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI64__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastI64,
  SafeCastI64Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI64";

const _abi = [
  {
    inputs: [],
    name: "OverflowInt64ToUint64",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212205f55a6a382453e5a4691ce40a9953fee79f8173f986303de98b04670f4c5ed7664736f6c63430008160033";

type SafeCastI64ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastI64ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastI64__factory extends ContractFactory {
  constructor(...args: SafeCastI64ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastI64> {
    return super.deploy(overrides || {}) as Promise<SafeCastI64>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastI64 {
    return super.attach(address) as SafeCastI64;
  }
  override connect(signer: Signer): SafeCastI64__factory {
    return super.connect(signer) as SafeCastI64__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastI64Interface {
    return new utils.Interface(_abi) as SafeCastI64Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastI64 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastI64;
  }
}
