// @synthetixio/perps-market/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI32__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastI32,
  SafeCastI32Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastI32";

const _abi = [
  {
    inputs: [],
    name: "OverflowInt32ToUint32",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122006eaf16a7ec22654cc756caaf446bbd4fcf1e2a76d953331b2aec1571256c1b764736f6c63430008160033";

type SafeCastI32ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastI32ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastI32__factory extends ContractFactory {
  constructor(...args: SafeCastI32ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastI32> {
    return super.deploy(overrides || {}) as Promise<SafeCastI32>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastI32 {
    return super.attach(address) as SafeCastI32;
  }
  override connect(signer: Signer): SafeCastI32__factory {
    return super.connect(signer) as SafeCastI32__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastI32Interface {
    return new utils.Interface(_abi) as SafeCastI32Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastI32 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastI32;
  }
}
