// @synthetixio/spot-market/test/generated/typechain/factories/MockPythERC7412Wrapper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  MockPythERC7412Wrapper,
  MockPythERC7412WrapperInterface,
} from "../MockPythERC7412Wrapper";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "priceId",
        type: "bytes32",
      },
      {
        internalType: "uint64",
        name: "requestedTime",
        type: "uint64",
      },
    ],
    name: "OracleDataRequired",
    type: "error",
  },
  {
    inputs: [],
    name: "alwaysRevert",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "priceId",
        type: "bytes32",
      },
      {
        internalType: "uint64",
        name: "requestedTime",
        type: "uint64",
      },
    ],
    name: "getBenchmarkPrice",
    outputs: [
      {
        internalType: "int256",
        name: "",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "price",
    outputs: [
      {
        internalType: "int256",
        name: "",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bool",
        name: "_alwaysRevert",
        type: "bool",
      },
    ],
    name: "setAlwaysRevertFlag",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "int256",
        name: "_price",
        type: "int256",
      },
    ],
    name: "setBenchmarkPrice",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class MockPythERC7412Wrapper__factory {
  static readonly abi = _abi;
  static createInterface(): MockPythERC7412WrapperInterface {
    return new utils.Interface(_abi) as MockPythERC7412WrapperInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): MockPythERC7412Wrapper {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as MockPythERC7412Wrapper;
  }
}
