// @lerna-lite/core/dist/prompt.js
import inquirer from 'inquirer';
import log from 'npmlog';
export async function promptConfirmation(message) {
    log.pause();
    const answers = await inquirer.prompt([
        {
            type: 'expand',
            name: 'confirm',
            message,
            default: 2,
            choices: [
                { key: 'y', name: 'Yes', value: true },
                { key: 'n', name: 'No', value: false },
            ],
        },
    ]);
    log.resume();
    return answers.confirm;
}
export async function promptSelectOne(message, { choices, filter, validate } = {}) {
    log.pause();
    const answers = await inquirer.prompt([
        {
            type: 'list',
            name: 'prompt',
            message,
            choices,
            pageSize: choices.length,
            filter,
            validate,
        },
    ]);
    log.resume();
    return answers.prompt;
}
export async function promptTextInput(message, { filter, validate } = {}) {
    log.pause();
    const answers = await inquirer.prompt([
        {
            type: 'input',
            name: 'input',
            message,
            filter,
            validate,
        },
    ]);
    log.resume();
    return answers.input;
}
