// @synthetixio/perps-market/typechain-types/contracts/mocks/MockGasPriceNode.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export declare namespace NodeDefinition {
  export type DataStruct = {
    nodeType: BigNumberish;
    parameters: BytesLike;
    parents: BytesLike[];
  };

  export type DataStructOutput = [number, string, string[]] & {
    nodeType: number;
    parameters: string;
    parents: string[];
  };
}

export declare namespace NodeOutput {
  export type DataStruct = {
    price: BigNumberish;
    timestamp: BigNumberish;
    __slotAvailableForFutureUse1: BigNumberish;
    __slotAvailableForFutureUse2: BigNumberish;
  };

  export type DataStructOutput = [
    BigNumber,
    BigNumber,
    BigNumber,
    BigNumber
  ] & {
    price: BigNumber;
    timestamp: BigNumber;
    __slotAvailableForFutureUse1: BigNumber;
    __slotAvailableForFutureUse2: BigNumber;
  };
}

export interface MockGasPriceNodeInterface extends utils.Interface {
  functions: {
    "KIND_FLAG()": FunctionFragment;
    "KIND_LIQUIDATE()": FunctionFragment;
    "KIND_SETTLEMENT()": FunctionFragment;
    "flagCost()": FunctionFragment;
    "isValid((uint8,bytes,bytes32[]))": FunctionFragment;
    "liquidateCost()": FunctionFragment;
    "process((int256,uint256,uint256,uint256)[],bytes,bytes32[],bytes32[])": FunctionFragment;
    "setCosts(uint256,uint256,uint256)": FunctionFragment;
    "settlementCost()": FunctionFragment;
    "supportsInterface(bytes4)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "KIND_FLAG"
      | "KIND_LIQUIDATE"
      | "KIND_SETTLEMENT"
      | "flagCost"
      | "isValid"
      | "liquidateCost"
      | "process"
      | "setCosts"
      | "settlementCost"
      | "supportsInterface"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "KIND_FLAG", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "KIND_LIQUIDATE",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "KIND_SETTLEMENT",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "flagCost", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "isValid",
    values: [NodeDefinition.DataStruct]
  ): string;
  encodeFunctionData(
    functionFragment: "liquidateCost",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "process",
    values: [NodeOutput.DataStruct[], BytesLike, BytesLike[], BytesLike[]]
  ): string;
  encodeFunctionData(
    functionFragment: "setCosts",
    values: [BigNumberish, BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "settlementCost",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "supportsInterface",
    values: [BytesLike]
  ): string;

  decodeFunctionResult(functionFragment: "KIND_FLAG", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "KIND_LIQUIDATE",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "KIND_SETTLEMENT",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "flagCost", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "isValid", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "liquidateCost",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "process", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "setCosts", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "settlementCost",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "supportsInterface",
    data: BytesLike
  ): Result;

  events: {};
}

export interface MockGasPriceNode extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: MockGasPriceNodeInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    KIND_FLAG(overrides?: CallOverrides): Promise<[BigNumber]>;

    KIND_LIQUIDATE(overrides?: CallOverrides): Promise<[BigNumber]>;

    KIND_SETTLEMENT(overrides?: CallOverrides): Promise<[BigNumber]>;

    flagCost(overrides?: CallOverrides): Promise<[BigNumber]>;

    isValid(
      nodeDefinition: NodeDefinition.DataStruct,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    liquidateCost(overrides?: CallOverrides): Promise<[BigNumber]>;

    process(
      arg0: NodeOutput.DataStruct[],
      arg1: BytesLike,
      runtimeKeys: BytesLike[],
      runtimeValues: BytesLike[],
      overrides?: CallOverrides
    ): Promise<[NodeOutput.DataStructOutput]>;

    setCosts(
      _settlementCost: BigNumberish,
      _flagCost: BigNumberish,
      _liquidateCost: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    settlementCost(overrides?: CallOverrides): Promise<[BigNumber]>;

    supportsInterface(
      arg0: BytesLike,
      overrides?: CallOverrides
    ): Promise<[boolean]>;
  };

  KIND_FLAG(overrides?: CallOverrides): Promise<BigNumber>;

  KIND_LIQUIDATE(overrides?: CallOverrides): Promise<BigNumber>;

  KIND_SETTLEMENT(overrides?: CallOverrides): Promise<BigNumber>;

  flagCost(overrides?: CallOverrides): Promise<BigNumber>;

  isValid(
    nodeDefinition: NodeDefinition.DataStruct,
    overrides?: CallOverrides
  ): Promise<boolean>;

  liquidateCost(overrides?: CallOverrides): Promise<BigNumber>;

  process(
    arg0: NodeOutput.DataStruct[],
    arg1: BytesLike,
    runtimeKeys: BytesLike[],
    runtimeValues: BytesLike[],
    overrides?: CallOverrides
  ): Promise<NodeOutput.DataStructOutput>;

  setCosts(
    _settlementCost: BigNumberish,
    _flagCost: BigNumberish,
    _liquidateCost: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  settlementCost(overrides?: CallOverrides): Promise<BigNumber>;

  supportsInterface(
    arg0: BytesLike,
    overrides?: CallOverrides
  ): Promise<boolean>;

  callStatic: {
    KIND_FLAG(overrides?: CallOverrides): Promise<BigNumber>;

    KIND_LIQUIDATE(overrides?: CallOverrides): Promise<BigNumber>;

    KIND_SETTLEMENT(overrides?: CallOverrides): Promise<BigNumber>;

    flagCost(overrides?: CallOverrides): Promise<BigNumber>;

    isValid(
      nodeDefinition: NodeDefinition.DataStruct,
      overrides?: CallOverrides
    ): Promise<boolean>;

    liquidateCost(overrides?: CallOverrides): Promise<BigNumber>;

    process(
      arg0: NodeOutput.DataStruct[],
      arg1: BytesLike,
      runtimeKeys: BytesLike[],
      runtimeValues: BytesLike[],
      overrides?: CallOverrides
    ): Promise<NodeOutput.DataStructOutput>;

    setCosts(
      _settlementCost: BigNumberish,
      _flagCost: BigNumberish,
      _liquidateCost: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    settlementCost(overrides?: CallOverrides): Promise<BigNumber>;

    supportsInterface(
      arg0: BytesLike,
      overrides?: CallOverrides
    ): Promise<boolean>;
  };

  filters: {};

  estimateGas: {
    KIND_FLAG(overrides?: CallOverrides): Promise<BigNumber>;

    KIND_LIQUIDATE(overrides?: CallOverrides): Promise<BigNumber>;

    KIND_SETTLEMENT(overrides?: CallOverrides): Promise<BigNumber>;

    flagCost(overrides?: CallOverrides): Promise<BigNumber>;

    isValid(
      nodeDefinition: NodeDefinition.DataStruct,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    liquidateCost(overrides?: CallOverrides): Promise<BigNumber>;

    process(
      arg0: NodeOutput.DataStruct[],
      arg1: BytesLike,
      runtimeKeys: BytesLike[],
      runtimeValues: BytesLike[],
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    setCosts(
      _settlementCost: BigNumberish,
      _flagCost: BigNumberish,
      _liquidateCost: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    settlementCost(overrides?: CallOverrides): Promise<BigNumber>;

    supportsInterface(
      arg0: BytesLike,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    KIND_FLAG(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    KIND_LIQUIDATE(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    KIND_SETTLEMENT(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    flagCost(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    isValid(
      nodeDefinition: NodeDefinition.DataStruct,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    liquidateCost(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    process(
      arg0: NodeOutput.DataStruct[],
      arg1: BytesLike,
      runtimeKeys: BytesLike[],
      runtimeValues: BytesLike[],
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    setCosts(
      _settlementCost: BigNumberish,
      _flagCost: BigNumberish,
      _liquidateCost: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    settlementCost(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    supportsInterface(
      arg0: BytesLike,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
