// @synthetixio/perps-market/typechain-types/factories/@synthetixio/main/contracts/storage/AccountRBAC__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  AccountRBAC,
  AccountRBACInterface,
} from "../../../../../@synthetixio/main/contracts/storage/AccountRBAC";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "permission",
        type: "bytes32",
      },
    ],
    name: "InvalidPermission",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220f040d1e0304e139324f1babb453df52e8fbcf1dd1b7b033d494453eac1a846af64736f6c63430008160033";

type AccountRBACConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AccountRBACConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AccountRBAC__factory extends ContractFactory {
  constructor(...args: AccountRBACConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<AccountRBAC> {
    return super.deploy(overrides || {}) as Promise<AccountRBAC>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): AccountRBAC {
    return super.attach(address) as AccountRBAC;
  }
  override connect(signer: Signer): AccountRBAC__factory {
    return super.connect(signer) as AccountRBAC__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AccountRBACInterface {
    return new utils.Interface(_abi) as AccountRBACInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AccountRBAC {
    return new Contract(address, _abi, signerOrProvider) as AccountRBAC;
  }
}
