// @synthetixio/perps-market/typechain-types/factories/@synthetixio/main/contracts/storage/Collateral__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  Collateral,
  CollateralInterface,
} from "../../../../../@synthetixio/main/contracts/storage/Collateral";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256",
        name: "amountAvailableForDelegationD18",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "amountD18",
        type: "uint256",
      },
    ],
    name: "InsufficentAvailableCollateral",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212207b3449209a2ec7e5e5aa11510857a3216542b99c4cc0f34f41ab81977aa7491464736f6c63430008160033";

type CollateralConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: CollateralConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class Collateral__factory extends ContractFactory {
  constructor(...args: CollateralConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<Collateral> {
    return super.deploy(overrides || {}) as Promise<Collateral>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): Collateral {
    return super.attach(address) as Collateral;
  }
  override connect(signer: Signer): Collateral__factory {
    return super.connect(signer) as Collateral__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): CollateralInterface {
    return new utils.Interface(_abi) as CollateralInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Collateral {
    return new Contract(address, _abi, signerOrProvider) as Collateral;
  }
}
