// @synthetixio/perps-market/typechain-types/factories/@synthetixio/main/contracts/storage/Distribution__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  Distribution,
  DistributionInterface,
} from "../../../../../@synthetixio/main/contracts/storage/Distribution";

const _abi = [
  {
    inputs: [],
    name: "EmptyDistribution",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122099b09b100fbf6dcb86d8c7d9909239b20eedb3099b720a9fe6cb5bfd44ccd07b64736f6c63430008160033";

type DistributionConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: DistributionConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class Distribution__factory extends ContractFactory {
  constructor(...args: DistributionConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<Distribution> {
    return super.deploy(overrides || {}) as Promise<Distribution>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): Distribution {
    return super.attach(address) as Distribution;
  }
  override connect(signer: Signer): Distribution__factory {
    return super.connect(signer) as Distribution__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): DistributionInterface {
    return new utils.Interface(_abi) as DistributionInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Distribution {
    return new Contract(address, _abi, signerOrProvider) as Distribution;
  }
}
