// @synthetixio/perps-market/typechain-types/factories/@synthetixio/main/contracts/storage/ScalableMapping__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ScalableMapping,
  ScalableMappingInterface,
} from "../../../../../@synthetixio/main/contracts/storage/ScalableMapping";

const _abi = [
  {
    inputs: [],
    name: "CannotScaleEmptyMapping",
    type: "error",
  },
  {
    inputs: [],
    name: "InsufficientMappedAmount",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212205c20f192ed61d69ddb52f141246831bea88539589a5fdc324be226becbda1f0364736f6c63430008160033";

type ScalableMappingConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ScalableMappingConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ScalableMapping__factory extends ContractFactory {
  constructor(...args: ScalableMappingConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ScalableMapping> {
    return super.deploy(overrides || {}) as Promise<ScalableMapping>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ScalableMapping {
    return super.attach(address) as ScalableMapping;
  }
  override connect(signer: Signer): ScalableMapping__factory {
    return super.connect(signer) as ScalableMapping__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ScalableMappingInterface {
    return new utils.Interface(_abi) as ScalableMappingInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ScalableMapping {
    return new Contract(address, _abi, signerOrProvider) as ScalableMapping;
  }
}
