// @synthetixio/spot-market/typechain-types/contracts/mocks/MockPythERC7412Wrapper.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface MockPythERC7412WrapperInterface extends utils.Interface {
  functions: {
    "alwaysRevert()": FunctionFragment;
    "getBenchmarkPrice(bytes32,uint64)": FunctionFragment;
    "price()": FunctionFragment;
    "setAlwaysRevertFlag(bool)": FunctionFragment;
    "setBenchmarkPrice(int256)": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "alwaysRevert"
      | "getBenchmarkPrice"
      | "price"
      | "setAlwaysRevertFlag"
      | "setBenchmarkPrice"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "alwaysRevert",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getBenchmarkPrice",
    values: [BytesLike, BigNumberish]
  ): string;
  encodeFunctionData(functionFragment: "price", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "setAlwaysRevertFlag",
    values: [boolean]
  ): string;
  encodeFunctionData(
    functionFragment: "setBenchmarkPrice",
    values: [BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "alwaysRevert",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getBenchmarkPrice",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "price", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "setAlwaysRevertFlag",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setBenchmarkPrice",
    data: BytesLike
  ): Result;

  events: {};
}

export interface MockPythERC7412Wrapper extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: MockPythERC7412WrapperInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    alwaysRevert(overrides?: CallOverrides): Promise<[boolean]>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    price(overrides?: CallOverrides): Promise<[BigNumber]>;

    setAlwaysRevertFlag(
      _alwaysRevert: boolean,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    setBenchmarkPrice(
      _price: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;
  };

  alwaysRevert(overrides?: CallOverrides): Promise<boolean>;

  getBenchmarkPrice(
    priceId: BytesLike,
    requestedTime: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  price(overrides?: CallOverrides): Promise<BigNumber>;

  setAlwaysRevertFlag(
    _alwaysRevert: boolean,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  setBenchmarkPrice(
    _price: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  callStatic: {
    alwaysRevert(overrides?: CallOverrides): Promise<boolean>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    price(overrides?: CallOverrides): Promise<BigNumber>;

    setAlwaysRevertFlag(
      _alwaysRevert: boolean,
      overrides?: CallOverrides
    ): Promise<void>;

    setBenchmarkPrice(
      _price: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    alwaysRevert(overrides?: CallOverrides): Promise<BigNumber>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    price(overrides?: CallOverrides): Promise<BigNumber>;

    setAlwaysRevertFlag(
      _alwaysRevert: boolean,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    setBenchmarkPrice(
      _price: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    alwaysRevert(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    price(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    setAlwaysRevertFlag(
      _alwaysRevert: boolean,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    setBenchmarkPrice(
      _price: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;
  };
}
