// @synthetixio/core-modules/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastU256__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastU256,
  SafeCastU256Interface,
} from "../../../../../../@synthetixio/core-contracts/contracts/utils/SafeCast/SafeCastU256";

const _abi = [
  {
    inputs: [],
    name: "OverflowUint256ToInt256",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint256ToUint128",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint256ToUint160",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint256ToUint32",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint256ToUint64",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220f012de3806430752f01c8ab0e1823ce8fad56a294ac842d792af08ef6bc7937a64736f6c63430008160033";

type SafeCastU256ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastU256ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastU256__factory extends ContractFactory {
  constructor(...args: SafeCastU256ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastU256> {
    return super.deploy(overrides || {}) as Promise<SafeCastU256>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastU256 {
    return super.attach(address) as SafeCastU256;
  }
  override connect(signer: Signer): SafeCastU256__factory {
    return super.connect(signer) as SafeCastU256__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastU256Interface {
    return new utils.Interface(_abi) as SafeCastU256Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastU256 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastU256;
  }
}
