// @synthetixio/spot-market/typechain-types/factories/@synthetixio/oracle-manager/contracts/nodes/StalenessCircuitBreakerNode__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  StalenessCircuitBreakerNode,
  StalenessCircuitBreakerNodeInterface,
} from "../../../../../@synthetixio/oracle-manager/contracts/nodes/StalenessCircuitBreakerNode";

const _abi = [
  {
    inputs: [],
    name: "StalenessToleranceExceeded",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212200c9cbd002ada3679a7b878739ab24b6323bc100a618ba69ae5d2ed74630b6db564736f6c63430008160033";

type StalenessCircuitBreakerNodeConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: StalenessCircuitBreakerNodeConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class StalenessCircuitBreakerNode__factory extends ContractFactory {
  constructor(...args: StalenessCircuitBreakerNodeConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<StalenessCircuitBreakerNode> {
    return super.deploy(
      overrides || {}
    ) as Promise<StalenessCircuitBreakerNode>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): StalenessCircuitBreakerNode {
    return super.attach(address) as StalenessCircuitBreakerNode;
  }
  override connect(signer: Signer): StalenessCircuitBreakerNode__factory {
    return super.connect(signer) as StalenessCircuitBreakerNode__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): StalenessCircuitBreakerNodeInterface {
    return new utils.Interface(_abi) as StalenessCircuitBreakerNodeInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): StalenessCircuitBreakerNode {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as StalenessCircuitBreakerNode;
  }
}
