// @synthetixio/core-modules/typechain-types/factories/contracts/modules/mocks/GenericModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  GenericModule,
  GenericModuleInterface,
} from "../../../../contracts/modules/mocks/GenericModule";

const _abi = [
  {
    inputs: [],
    name: "getFortyTwo",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "pure",
    type: "function",
  },
] as const;

const _bytecode =
  "0x6080604052348015600f57600080fd5b50607780601d6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c80630690e16014602d575b600080fd5b602a60405190815260200160405180910390f3fea2646970667358221220169644594ad6aa800f526b5c04430a436551cac739883783d0289292c95ba16d64736f6c63430008160033";

type GenericModuleConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: GenericModuleConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class GenericModule__factory extends ContractFactory {
  constructor(...args: GenericModuleConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<GenericModule> {
    return super.deploy(overrides || {}) as Promise<GenericModule>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): GenericModule {
    return super.attach(address) as GenericModule;
  }
  override connect(signer: Signer): GenericModule__factory {
    return super.connect(signer) as GenericModule__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): GenericModuleInterface {
    return new utils.Interface(_abi) as GenericModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): GenericModule {
    return new Contract(address, _abi, signerOrProvider) as GenericModule;
  }
}
