// @synthetixio/core-modules/typechain-types/factories/contracts/storage/FeatureFlag__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  FeatureFlag,
  FeatureFlagInterface,
} from "../../../contracts/storage/FeatureFlag";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "which",
        type: "bytes32",
      },
    ],
    name: "FeatureUnavailable",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220c7720eca1fd7de4266b74838a7ec506d3ac2b70e0288ebdc248517e8ac3d4dbc64736f6c63430008160033";

type FeatureFlagConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: FeatureFlagConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class FeatureFlag__factory extends ContractFactory {
  constructor(...args: FeatureFlagConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<FeatureFlag> {
    return super.deploy(overrides || {}) as Promise<FeatureFlag>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): FeatureFlag {
    return super.attach(address) as FeatureFlag;
  }
  override connect(signer: Signer): FeatureFlag__factory {
    return super.connect(signer) as FeatureFlag__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): FeatureFlagInterface {
    return new utils.Interface(_abi) as FeatureFlagInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): FeatureFlag {
    return new Contract(address, _abi, signerOrProvider) as FeatureFlag;
  }
}
