// @synthetixio/perps-market/typechain-types/contracts/interfaces/external/IPythERC7412Wrapper.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../../common";

export interface IPythERC7412WrapperInterface extends utils.Interface {
  functions: {
    "getBenchmarkPrice(bytes32,uint64)": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "getBenchmarkPrice"): FunctionFragment;

  encodeFunctionData(
    functionFragment: "getBenchmarkPrice",
    values: [BytesLike, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "getBenchmarkPrice",
    data: BytesLike
  ): Result;

  events: {};
}

export interface IPythERC7412Wrapper extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IPythERC7412WrapperInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;
  };

  getBenchmarkPrice(
    priceId: BytesLike,
    requestedTime: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  callStatic: {
    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  filters: {};

  estimateGas: {
    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    getBenchmarkPrice(
      priceId: BytesLike,
      requestedTime: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
