// @synthetixio/perps-market/typechain-types/factories/@synthetixio/oracle-manager/contracts/nodes/ChainlinkNode__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ChainlinkNode,
  ChainlinkNodeInterface,
} from "../../../../../@synthetixio/oracle-manager/contracts/nodes/ChainlinkNode";

const _abi = [
  {
    inputs: [],
    name: "PRECISION",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

const _bytecode =
  "0x6087610038600b82828239805160001a607314602b57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe730000000000000000000000000000000000000000301460806040526004361060335760003560e01c8063aaf5eb68146038575b600080fd5b603f601281565b60405190815260200160405180910390f3fea2646970667358221220cf9c2b180bebad923fe18ba62f876fd4815712bbd2b36f2462437ca092373ad064736f6c63430008160033";

type ChainlinkNodeConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ChainlinkNodeConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ChainlinkNode__factory extends ContractFactory {
  constructor(...args: ChainlinkNodeConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ChainlinkNode> {
    return super.deploy(overrides || {}) as Promise<ChainlinkNode>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ChainlinkNode {
    return super.attach(address) as ChainlinkNode;
  }
  override connect(signer: Signer): ChainlinkNode__factory {
    return super.connect(signer) as ChainlinkNode__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ChainlinkNodeInterface {
    return new utils.Interface(_abi) as ChainlinkNodeInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ChainlinkNode {
    return new Contract(address, _abi, signerOrProvider) as ChainlinkNode;
  }
}
