// @synthetixio/perps-market/typechain-types/factories/@synthetixio/oracle-manager/contracts/nodes/ReducerNode__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ReducerNode,
  ReducerNodeInterface,
} from "../../../../../@synthetixio/oracle-manager/contracts/nodes/ReducerNode";

const _abi = [
  {
    inputs: [
      {
        internalType: "int256",
        name: "price",
        type: "int256",
      },
    ],
    name: "InvalidPrice",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "enum ReducerNode.Operations",
        name: "operation",
        type: "uint8",
      },
    ],
    name: "UnsupportedOperation",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220e150096c5a3966509b3fa32fbadf280e347db46b46aff9d5f71cec8148813fd264736f6c63430008160033";

type ReducerNodeConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ReducerNodeConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ReducerNode__factory extends ContractFactory {
  constructor(...args: ReducerNodeConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ReducerNode> {
    return super.deploy(overrides || {}) as Promise<ReducerNode>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ReducerNode {
    return super.attach(address) as ReducerNode;
  }
  override connect(signer: Signer): ReducerNode__factory {
    return super.connect(signer) as ReducerNode__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ReducerNodeInterface {
    return new utils.Interface(_abi) as ReducerNodeInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ReducerNode {
    return new Contract(address, _abi, signerOrProvider) as ReducerNode;
  }
}
