// @synthetixio/perps-market/test/generated/typechain/factories/PerpsAccountModule__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  PerpsAccountModule,
  PerpsAccountModuleInterface,
} from "../PerpsAccountModule";

const _abi = [
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "AccountLiquidatable",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "AccountNotFound",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "which",
        type: "bytes32",
      },
    ],
    name: "FeatureUnavailable",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "collateralAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "withdrawAmount",
        type: "uint256",
      },
    ],
    name: "InsufficientCollateral",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "int256",
        name: "withdrawableMarginUsd",
        type: "int256",
      },
      {
        internalType: "uint256",
        name: "requestedMarginUsd",
        type: "uint256",
      },
    ],
    name: "InsufficientCollateralAvailableForWithdraw",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "collateralAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "withdrawAmount",
        type: "uint256",
      },
    ],
    name: "InsufficientSynthCollateral",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "int256",
        name: "amountDelta",
        type: "int256",
      },
    ],
    name: "InvalidAmountDelta",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "id",
        type: "uint128",
      },
    ],
    name: "InvalidId",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "InvalidMarket",
    type: "error",
  },
  {
    inputs: [],
    name: "KeeperCostsNotSet",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "maxAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "collateralAmount",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "depositAmount",
        type: "uint256",
      },
    ],
    name: "MaxCollateralExceeded",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "maxCollateralsPerAccount",
        type: "uint128",
      },
    ],
    name: "MaxCollateralsPerAccountReached",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "NonexistentDebt",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowInt256ToUint256",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint128ToInt128",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint256ToInt256",
    type: "error",
  },
  {
    inputs: [],
    name: "OverflowUint256ToUint128",
    type: "error",
  },
  {
    inputs: [],
    name: "PendingOrderExists",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        internalType: "bytes32",
        name: "permission",
        type: "bytes32",
      },
      {
        internalType: "address",
        name: "target",
        type: "address",
      },
    ],
    name: "PermissionDenied",
    type: "error",
  },
  {
    inputs: [],
    name: "PositionOutOfBounds",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "PriceFeedNotSet",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
    ],
    name: "SynthNotEnabledForCollateral",
    type: "error",
  },
  {
    inputs: [],
    name: "ValueAlreadyInSet",
    type: "error",
  },
  {
    inputs: [],
    name: "ValueNotInSet",
    type: "error",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        indexed: true,
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "int256",
        name: "amountDelta",
        type: "int256",
      },
      {
        indexed: true,
        internalType: "address",
        name: "sender",
        type: "address",
      },
    ],
    name: "CollateralModified",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
      {
        indexed: true,
        internalType: "address",
        name: "sender",
        type: "address",
      },
    ],
    name: "DebtPaid",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "debt",
    outputs: [
      {
        internalType: "uint256",
        name: "accountDebt",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "getAccountCollateralIds",
    outputs: [
      {
        internalType: "uint256[]",
        name: "",
        type: "uint256[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "getAccountOpenPositions",
    outputs: [
      {
        internalType: "uint256[]",
        name: "",
        type: "uint256[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "getAvailableMargin",
    outputs: [
      {
        internalType: "int256",
        name: "availableMargin",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
    ],
    name: "getCollateralAmount",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "getOpenPosition",
    outputs: [
      {
        internalType: "int256",
        name: "totalPnl",
        type: "int256",
      },
      {
        internalType: "int256",
        name: "accruedFunding",
        type: "int256",
      },
      {
        internalType: "int128",
        name: "positionSize",
        type: "int128",
      },
      {
        internalType: "uint256",
        name: "owedInterest",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "marketId",
        type: "uint128",
      },
    ],
    name: "getOpenPositionSize",
    outputs: [
      {
        internalType: "int128",
        name: "positionSize",
        type: "int128",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "getRequiredMargins",
    outputs: [
      {
        internalType: "uint256",
        name: "requiredInitialMargin",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "requiredMaintenanceMargin",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "maxLiquidationReward",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "getWithdrawableMargin",
    outputs: [
      {
        internalType: "int256",
        name: "withdrawableMargin",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        internalType: "uint128",
        name: "collateralId",
        type: "uint128",
      },
      {
        internalType: "int256",
        name: "amountDelta",
        type: "int256",
      },
    ],
    name: "modifyCollateral",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
      {
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "payDebt",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "totalAccountOpenInterest",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint128",
        name: "accountId",
        type: "uint128",
      },
    ],
    name: "totalCollateralValue",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class PerpsAccountModule__factory {
  static readonly abi = _abi;
  static createInterface(): PerpsAccountModuleInterface {
    return new utils.Interface(_abi) as PerpsAccountModuleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): PerpsAccountModule {
    return new Contract(address, _abi, signerOrProvider) as PerpsAccountModule;
  }
}
