// @synthetixio/perps-market/typechain-types/contracts/utils/Flags.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface FlagsInterface extends utils.Interface {
  functions: {
    "CREATE_MARKET()": FunctionFragment;
    "PERPS_SYSTEM()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "CREATE_MARKET" | "PERPS_SYSTEM"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "CREATE_MARKET",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "PERPS_SYSTEM",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "CREATE_MARKET",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "PERPS_SYSTEM",
    data: BytesLike
  ): Result;

  events: {};
}

export interface Flags extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: FlagsInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    CREATE_MARKET(overrides?: CallOverrides): Promise<[string]>;

    PERPS_SYSTEM(overrides?: CallOverrides): Promise<[string]>;
  };

  CREATE_MARKET(overrides?: CallOverrides): Promise<string>;

  PERPS_SYSTEM(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    CREATE_MARKET(overrides?: CallOverrides): Promise<string>;

    PERPS_SYSTEM(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    CREATE_MARKET(overrides?: CallOverrides): Promise<BigNumber>;

    PERPS_SYSTEM(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    CREATE_MARKET(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    PERPS_SYSTEM(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
