// @synthetixio/core-contracts/typechain-types/factories/utils/SafeCast/SafeCastI56__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastI56,
  SafeCastI56Interface,
} from "../../../utils/SafeCast/SafeCastI56";

const _abi = [
  {
    inputs: [],
    name: "OverflowInt56ToInt24",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212205bccb1d292323fcbe66db3966a29cb20f8684453db699e17cdde1d518712a9da64736f6c63430008160033";

type SafeCastI56ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastI56ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastI56__factory extends ContractFactory {
  constructor(...args: SafeCastI56ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastI56> {
    return super.deploy(overrides || {}) as Promise<SafeCastI56>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastI56 {
    return super.attach(address) as SafeCastI56;
  }
  override connect(signer: Signer): SafeCastI56__factory {
    return super.connect(signer) as SafeCastI56__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastI56Interface {
    return new utils.Interface(_abi) as SafeCastI56Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastI56 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastI56;
  }
}
