// @synthetixio/core-contracts/typechain-types/factories/utils/SafeCast/SafeCastU128__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SafeCastU128,
  SafeCastU128Interface,
} from "../../../utils/SafeCast/SafeCastU128";

const _abi = [
  {
    inputs: [],
    name: "OverflowUint128ToInt128",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212207f9e790a6b960b9450d5bfb20332dd4ba6998ce782eb0abb205f0ea67fdac0b064736f6c63430008160033";

type SafeCastU128ConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SafeCastU128ConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SafeCastU128__factory extends ContractFactory {
  constructor(...args: SafeCastU128ConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<SafeCastU128> {
    return super.deploy(overrides || {}) as Promise<SafeCastU128>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SafeCastU128 {
    return super.attach(address) as SafeCastU128;
  }
  override connect(signer: Signer): SafeCastU128__factory {
    return super.connect(signer) as SafeCastU128__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SafeCastU128Interface {
    return new utils.Interface(_abi) as SafeCastU128Interface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SafeCastU128 {
    return new Contract(address, _abi, signerOrProvider) as SafeCastU128;
  }
}
