// @synthetixio/perps-market/typechain-types/contracts/interfaces/IAccountEvents.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type { BaseContract, BigNumber, Signer, utils } from "ethers";
import type { EventFragment } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../common";

export interface IAccountEventsInterface extends utils.Interface {
  functions: {};

  events: {
    "AccountCharged(uint128,int256,uint256)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "AccountCharged"): EventFragment;
}

export interface AccountChargedEventObject {
  accountId: BigNumber;
  amount: BigNumber;
  accountDebt: BigNumber;
}
export type AccountChargedEvent = TypedEvent<
  [BigNumber, BigNumber, BigNumber],
  AccountChargedEventObject
>;

export type AccountChargedEventFilter = TypedEventFilter<AccountChargedEvent>;

export interface IAccountEvents extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IAccountEventsInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {};

  callStatic: {};

  filters: {
    "AccountCharged(uint128,int256,uint256)"(
      accountId?: null,
      amount?: null,
      accountDebt?: null
    ): AccountChargedEventFilter;
    AccountCharged(
      accountId?: null,
      amount?: null,
      accountDebt?: null
    ): AccountChargedEventFilter;
  };

  estimateGas: {};

  populateTransaction: {};
}
