// @synthetixio/perps-market/typechain-types/@synthetixio/core-modules/contracts/storage/AssociatedSystem.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "../../../../common";

export interface AssociatedSystemInterface extends utils.Interface {
  functions: {
    "KIND_ERC20()": FunctionFragment;
    "KIND_ERC721()": FunctionFragment;
    "KIND_UNMANAGED()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "KIND_ERC20" | "KIND_ERC721" | "KIND_UNMANAGED"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "KIND_ERC20",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "KIND_ERC721",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "KIND_UNMANAGED",
    values?: undefined
  ): string;

  decodeFunctionResult(functionFragment: "KIND_ERC20", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "KIND_ERC721",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "KIND_UNMANAGED",
    data: BytesLike
  ): Result;

  events: {};
}

export interface AssociatedSystem extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: AssociatedSystemInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    KIND_ERC20(overrides?: CallOverrides): Promise<[string]>;

    KIND_ERC721(overrides?: CallOverrides): Promise<[string]>;

    KIND_UNMANAGED(overrides?: CallOverrides): Promise<[string]>;
  };

  KIND_ERC20(overrides?: CallOverrides): Promise<string>;

  KIND_ERC721(overrides?: CallOverrides): Promise<string>;

  KIND_UNMANAGED(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    KIND_ERC20(overrides?: CallOverrides): Promise<string>;

    KIND_ERC721(overrides?: CallOverrides): Promise<string>;

    KIND_UNMANAGED(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    KIND_ERC20(overrides?: CallOverrides): Promise<BigNumber>;

    KIND_ERC721(overrides?: CallOverrides): Promise<BigNumber>;

    KIND_UNMANAGED(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    KIND_ERC20(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    KIND_ERC721(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    KIND_UNMANAGED(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
