// @synthetixio/core-contracts/typechain-types/factories/errors/AddressError__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  AddressError,
  AddressErrorInterface,
} from "../../errors/AddressError";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "contr",
        type: "address",
      },
    ],
    name: "NotAContract",
    type: "error",
  },
  {
    inputs: [],
    name: "ZeroAddress",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122038589952697cae60ac92c8db4ff0ce72a2d7c433f9cdaefaf3734577b142bdb464736f6c63430008160033";

type AddressErrorConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: AddressErrorConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class AddressError__factory extends ContractFactory {
  constructor(...args: AddressErrorConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<AddressError> {
    return super.deploy(overrides || {}) as Promise<AddressError>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): AddressError {
    return super.attach(address) as AddressError;
  }
  override connect(signer: Signer): AddressError__factory {
    return super.connect(signer) as AddressError__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): AddressErrorInterface {
    return new utils.Interface(_abi) as AddressErrorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): AddressError {
    return new Contract(address, _abi, signerOrProvider) as AddressError;
  }
}
