// @synthetixio/core-contracts/typechain-types/factories/errors/ArrayError__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type { ArrayError, ArrayErrorInterface } from "../../errors/ArrayError";

const _abi = [
  {
    inputs: [],
    name: "EmptyArray",
    type: "error",
  },
  {
    inputs: [],
    name: "OutOfBounds",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea26469706673582212207f6fcc6a92daacb792b5dce0d9805850863609d3486935dd213885d4c4058fcf64736f6c63430008160033";

type ArrayErrorConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ArrayErrorConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ArrayError__factory extends ContractFactory {
  constructor(...args: ArrayErrorConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ArrayError> {
    return super.deploy(overrides || {}) as Promise<ArrayError>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ArrayError {
    return super.attach(address) as ArrayError;
  }
  override connect(signer: Signer): ArrayError__factory {
    return super.connect(signer) as ArrayError__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ArrayErrorInterface {
    return new utils.Interface(_abi) as ArrayErrorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ArrayError {
    return new Contract(address, _abi, signerOrProvider) as ArrayError;
  }
}
