// @synthetixio/core-contracts/typechain-types/factories/errors/ChangeError__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  ChangeError,
  ChangeErrorInterface,
} from "../../errors/ChangeError";

const _abi = [
  {
    inputs: [],
    name: "NoChange",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea264697066735822122037e34c6c046bc4fae1b4dbcde0cc5bf6db984e472e41ce4a265436e08c4e094164736f6c63430008160033";

type ChangeErrorConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: ChangeErrorConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class ChangeError__factory extends ContractFactory {
  constructor(...args: ChangeErrorConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(
    overrides?: Overrides & { from?: string }
  ): Promise<ChangeError> {
    return super.deploy(overrides || {}) as Promise<ChangeError>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): ChangeError {
    return super.attach(address) as ChangeError;
  }
  override connect(signer: Signer): ChangeError__factory {
    return super.connect(signer) as ChangeError__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): ChangeErrorInterface {
    return new utils.Interface(_abi) as ChangeErrorInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ChangeError {
    return new Contract(address, _abi, signerOrProvider) as ChangeError;
  }
}
