// @synthetixio/core-modules/typechain-types/factories/@synthetixio/core-contracts/contracts/utils/SetUtil__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Signer, utils, Contract, ContractFactory, Overrides } from "ethers";
import type { Provider, TransactionRequest } from "@ethersproject/providers";
import type {
  SetUtil,
  SetUtilInterface,
} from "../../../../../@synthetixio/core-contracts/contracts/utils/SetUtil";

const _abi = [
  {
    inputs: [],
    name: "PositionOutOfBounds",
    type: "error",
  },
  {
    inputs: [],
    name: "ValueAlreadyInSet",
    type: "error",
  },
  {
    inputs: [],
    name: "ValueNotInSet",
    type: "error",
  },
] as const;

const _bytecode =
  "0x60566037600b82828239805160001a607314602a57634e487b7160e01b600052600060045260246000fd5b30600052607381538281f3fe73000000000000000000000000000000000000000030146080604052600080fdfea2646970667358221220f9f05fc719629a962998b20223e177481fa1a3564acdaa58e77519bcd9a063a464736f6c63430008160033";

type SetUtilConstructorParams =
  | [signer?: Signer]
  | ConstructorParameters<typeof ContractFactory>;

const isSuperArgs = (
  xs: SetUtilConstructorParams
): xs is ConstructorParameters<typeof ContractFactory> => xs.length > 1;

export class SetUtil__factory extends ContractFactory {
  constructor(...args: SetUtilConstructorParams) {
    if (isSuperArgs(args)) {
      super(...args);
    } else {
      super(_abi, _bytecode, args[0]);
    }
  }

  override deploy(overrides?: Overrides & { from?: string }): Promise<SetUtil> {
    return super.deploy(overrides || {}) as Promise<SetUtil>;
  }
  override getDeployTransaction(
    overrides?: Overrides & { from?: string }
  ): TransactionRequest {
    return super.getDeployTransaction(overrides || {});
  }
  override attach(address: string): SetUtil {
    return super.attach(address) as SetUtil;
  }
  override connect(signer: Signer): SetUtil__factory {
    return super.connect(signer) as SetUtil__factory;
  }

  static readonly bytecode = _bytecode;
  static readonly abi = _abi;
  static createInterface(): SetUtilInterface {
    return new utils.Interface(_abi) as SetUtilInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): SetUtil {
    return new Contract(address, _abi, signerOrProvider) as SetUtil;
  }
}
