// @synthetixio/spot-market/typechain-types/factories/contracts/interfaces/external/IChainlinkVerifier__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IChainlinkVerifier,
  IChainlinkVerifierInterface,
} from "../../../../contracts/interfaces/external/IChainlinkVerifier";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes",
        name: "chainlinkBlob",
        type: "bytes",
      },
    ],
    name: "verify",
    outputs: [
      {
        internalType: "bytes",
        name: "verifierResponse",
        type: "bytes",
      },
    ],
    stateMutability: "nonpayable",
    type: "function",
  },
] as const;

export class IChainlinkVerifier__factory {
  static readonly abi = _abi;
  static createInterface(): IChainlinkVerifierInterface {
    return new utils.Interface(_abi) as IChainlinkVerifierInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IChainlinkVerifier {
    return new Contract(address, _abi, signerOrProvider) as IChainlinkVerifier;
  }
}
