// @synthetixio/spot-market/typechain-types/factories/contracts/interfaces/external/IPythERC7412Wrapper__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IPythERC7412Wrapper,
  IPythERC7412WrapperInterface,
} from "../../../../contracts/interfaces/external/IPythERC7412Wrapper";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "oracleContract",
        type: "address",
      },
      {
        internalType: "bytes",
        name: "oracleQuery",
        type: "bytes",
      },
    ],
    name: "OracleDataRequired",
    type: "error",
  },
  {
    inputs: [
      {
        internalType: "bytes",
        name: "signedOffchainData",
        type: "bytes",
      },
    ],
    name: "fulfillOracleQuery",
    outputs: [],
    stateMutability: "payable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "priceId",
        type: "bytes32",
      },
      {
        internalType: "uint64",
        name: "requestedTime",
        type: "uint64",
      },
    ],
    name: "getBenchmarkPrice",
    outputs: [
      {
        internalType: "int256",
        name: "",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "priceId",
        type: "bytes32",
      },
      {
        internalType: "uint256",
        name: "stalenessTolerance",
        type: "uint256",
      },
    ],
    name: "getLatestPrice",
    outputs: [
      {
        internalType: "int256",
        name: "",
        type: "int256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class IPythERC7412Wrapper__factory {
  static readonly abi = _abi;
  static createInterface(): IPythERC7412WrapperInterface {
    return new utils.Interface(_abi) as IPythERC7412WrapperInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPythERC7412Wrapper {
    return new Contract(address, _abi, signerOrProvider) as IPythERC7412Wrapper;
  }
}
